{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant calling analysis Killswitch circuit \n",
    "\n",
    "Defining variants (SNPs, insertions and deletions) for a set of HiSeq samples exploring the variability in populations of *M. pneumoniae* transformed with a killswitch cassette (C5 samples) or two (C20 samples); sequencing performed at different selection passages (2, 3, 15; 2 replicates in each condition and cassette) and a pair of replicates at passage 3 were the cassette was induced by IPTG.\n",
    "\n",
    "**Authors:** [Samuel Miravet-Verde](mailto:samuel.miravet@crg.eu) and [Alicia Broto](mailto:alicia.broto@crg.eu) \n",
    "\n",
    "**Last update:** 14/06/2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib widget\n",
    "\n",
    "import sys, os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scripts\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General info about the genome references and the location of the cassettes in the genomes\n",
    "seq_record5 = SeqIO.read('./data/C5_refSeq.gb', \"genbank\")\n",
    "seq_record20 = SeqIO.read('./data/C20_refSeq.gb', \"genbank\")\n",
    "genome_C5 = str(seq_record5.seq)\n",
    "genome_C20 = str(seq_record5.seq)\n",
    "\n",
    "cass_par = [565510, 572124] # Present in C5 and C20\n",
    "cass_3b = [780850, 787445]  # Only in C20\n",
    "\n",
    "# Annotation information to work faster with dataframes\n",
    "ncbi = pd.read_csv('./data/mpn_annotation.csv', sep='\\t', header=None) # Genome annotation\n",
    "ncbi.columns = ['gene', 'start', 'end', 'strand']\n",
    "ncbi.set_index('gene', inplace=True)\n",
    "gold = pd.read_csv('./data/goldsets.csv', sep='\\t')  # Gold set \n",
    "\n",
    "posN = []\n",
    "posE = []\n",
    "for gene, cat in zip(gold['gene'], gold['class']):\n",
    "    if cat=='E':\n",
    "        posE += range(ncbi.loc[gene][0], ncbi.loc[gene][1]+1)\n",
    "    else:\n",
    "        posN += range(ncbi.loc[gene][0], ncbi.loc[gene][1]+1)\n",
    "        \n",
    "# Define the different set of positions in the genome were we will perform the analysis\n",
    "posN = set(posN)\n",
    "posE = set(posE)\n",
    "cas1 = set(range(cass_par[0], cass_par[1]+1))\n",
    "cas2 = set(range(cass_3b[0], cass_3b[1]+1))\n",
    "cas12 = cas1.union(cas2)\n",
    "geno = set(range(1, 816395)).difference(cas1.union(cas2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Map variations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will be using [Snippy](https://github.com/tseemann/snippy) for the first steps related to sequence mapping. To run it in parallel for all the samples we create a input file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/C20p3/B_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C20p3/B_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3/D_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3/D_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3IPTG/A_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3IPTG/A_S1b_L001_R1_001.fastq.gz',\n",
       " '../data/C20p15/7_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C20p15/7_S5_L001_R1_001.fastq.gz',\n",
       " '../data/C5p15/F_S5_L001_R1_001.fastq.gz',\n",
       " '../data/C5p15/F_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C20p2/6_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C20p2/6_S4_L001_R1_001.fastq.gz']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### 1. Make a multifile to map with snippy-multi.\n",
    "files = glob.glob('../data/C[5|20]*/*_R1_*') # The gre expression refers to the raw sequencing files located in data (this has to point to the directory where fastq files are stored)\n",
    "\n",
    "fo1 = open('./tmp_files/snippy_input1.txt', 'w')\n",
    "fo2 = open('./tmp_files/snippy_input2.txt', 'w')\n",
    "for f in files:\n",
    "    ide = f.split('/')[2:]\n",
    "    if 'C5' in f:\n",
    "        ide = ide[0].replace('C5', '')+'_'+'_'.join(ide[1].split('_')[:-3])\n",
    "        fo1.write('{}\\t{}\\t{}\\n'.format(ide, f, f.replace('_R1_', '_R2_')))\n",
    "    else:\n",
    "        ide = ide[0].replace('C20', '')+'_'+'_'.join(ide[1].split('_')[:-3])\n",
    "        fo2.write('{}\\t{}\\t{}\\n'.format(ide, f, f.replace('_R1_', '_R2_')))\n",
    "fo1.close()\n",
    "fo2.close()\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run the command in the snippy environment:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "conda activate snippy\n",
    "snippy-multi ./data/snippy_input.txt --ref ./data/C5_refSeq.gb --cpus 4 --report >  ./snippy_run.sh'\n",
    "sh ./snippy_run.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run freebayes in a permissive manner. This command will create a directory with the name of the sample were all the files will be stored. This is expected to be run in the terminal copy-pasting the printed command. It can also be run changing the print to `os.system` to execute it for each file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Listed samples in order\n",
    "order1 = ['p2_2_S1','p2_2_S2','p2_2_S3','p3_D_S2', 'p3_D_S3','p15_F_S1','p15_F_S5', 'p3IPTG_A_S1', 'p3IPTG_A_S1b']\n",
    "order2 = ['p2_6_S2','p2_6_S4','p3_B_S1','p3_B_S2','p15_7_S3','p15_7_S5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "freebayes-parallel p3_D_S2/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3_D_S2/reference/ref.fa ../p3_D_S2/snps.bam > filter1/p3_D_S2.raw.vcf\n",
      "freebayes-parallel p2_6_S2/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_6_S2/reference/ref.fa ../p2_6_S2/snps.bam > filter2/p2_6_S2.raw.vcf\n",
      "freebayes-parallel p2_2_S1/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_2_S1/reference/ref.fa ../p2_2_S1/snps.bam > filter1/p2_2_S1.raw.vcf\n",
      "freebayes-parallel p15_7_S5/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_7_S5/reference/ref.fa ../p15_7_S5/snps.bam > filter2/p15_7_S5.raw.vcf\n",
      "freebayes-parallel p3IPTG_A_S1/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3IPTG_A_S1/reference/ref.fa ../p3IPTG_A_S1/snps.bam > filter1/p3IPTG_A_S1.raw.vcf\n",
      "freebayes-parallel p3_D_S3/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3_D_S3/reference/ref.fa ../p3_D_S3/snps.bam > filter1/p3_D_S3.raw.vcf\n",
      "freebayes-parallel p3_B_S1/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3_B_S1/reference/ref.fa ../p3_B_S1/snps.bam > filter2/p3_B_S1.raw.vcf\n",
      "freebayes-parallel p15_7_S3/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_7_S3/reference/ref.fa ../p15_7_S3/snps.bam > filter2/p15_7_S3.raw.vcf\n",
      "freebayes-parallel p2_2_S3/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_2_S3/reference/ref.fa ../p2_2_S3/snps.bam > filter1/p2_2_S3.raw.vcf\n",
      "freebayes-parallel p15_F_S1/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_F_S1/reference/ref.fa ../p15_F_S1/snps.bam > filter1/p15_F_S1.raw.vcf\n",
      "freebayes-parallel p2_6_S4/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_6_S4/reference/ref.fa ../p2_6_S4/snps.bam > filter2/p2_6_S4.raw.vcf\n",
      "freebayes-parallel p2_2_S2/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_2_S2/reference/ref.fa ../p2_2_S2/snps.bam > filter1/p2_2_S2.raw.vcf\n",
      "freebayes-parallel p15_F_S5/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_F_S5/reference/ref.fa ../p15_F_S5/snps.bam > filter1/p15_F_S5.raw.vcf\n",
      "freebayes-parallel p3_B_S2/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3_B_S2/reference/ref.fa ../p3_B_S2/snps.bam > filter2/p3_B_S2.raw.vcf\n",
      "freebayes-parallel p3IPTG_A_S1b/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p3IPTG_A_S1b/reference/ref.fa ../p3IPTG_A_S1b/snps.bam > filter1/p3IPTG_A_S1b.raw.vcf\n"
     ]
    }
   ],
   "source": [
    "# freebayes-parallel reference/ref.txt 4 -p 2 -P 0 -C 2 -F 0.05 --min-coverage 10 --min-repeat-entropy 1.0 -q 13 -m 60 --strict-vcf   -f reference/ref.fa snps.bam > snps.raw.vcf\n",
    "for fil in glob.glob('../p*/*.bam'):\n",
    "    # again this has to point to directories were snipy output is stored\n",
    "    ide = fil.split('/')[1]\n",
    "    #print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.01 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter10/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    #print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 2 -F 0.05 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter50/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    if ide in order1:\n",
    "        print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter1/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    else:\n",
    "        print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter2/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get the total number of mapped reads:\n",
    "\n",
    "```python\n",
    "for fil in glob.glob('./p*/*.bam'):\n",
    "    ide = fil.split('/')[1]\n",
    "    # get the total number of reads of a BAM file (may include unmapped and duplicated multi-aligned reads)\n",
    "    print('samtools view -c {} >> total_readcount.txt'.format(fil))\n",
    "    print('samtools view -c {} -F 260 >> total_readcount_mapped.txt'.format(fil))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally we annotate the effect of the variants using SnpEff. This requires to add the custom annotation to the sources, specifying that the translation code is set to 4 for *M. pneumoniae*.\n",
    "\n",
    "1. Download SnpEff and edited the config (in software_crg/SnpEff) to include the C5 and C20 refseq references\n",
    "2. Build the custom db:\n",
    "```bash\n",
    "java -jar snpEff.jar build -genbank -v C5ali\n",
    "java -jar snpEff.jar build -genbank -v C20ali\n",
    "```\n",
    "3. Annotate vcf files in bash (in directory of the software)\n",
    "```bash\n",
    "for fil in `ls ./filter1/p*.vcf`; do java -Xmx8g -jar snpEff.jar C5ali $fil > $fil.ann; done\n",
    "for fil in `ls ./filter2/p*.vcf`; do java -Xmx8g -jar snpEff.jar C20ali $fil > $fil.ann; done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Data loading \n",
    "\n",
    "We will parse the variations to keep the columns of interest to perform the exploratory analysis. Dataframes include the following information (by column header):\n",
    "- SAMPLE: sample identifier includes information about the passage (e.g. p3IPTG is the passage 3 induced)\n",
    "- PASS: passage (2,3 or 15)\n",
    "- POS: genome loci (base pair position) where the variant is found\n",
    "- QUAL: estimate of the probability that there is a polymorphism at the loci described by the record\n",
    "- TOT: total number of reads covering a loci\n",
    "- REFN: total number of reads covering a loci matching the reference\n",
    "- ALTN: number of reads presenting a variant\n",
    "- REF: reference sequence\n",
    "- ALT: alternative sequence\n",
    "- EFF: potential effect of the variant as given by snpeff\n",
    "- IMPACT: potential impact of the variant, can be LOW (synonimous mutations), MODERATE (missense), HIGH (non-synonimous mutations, start lost, stop lost), or MODIFIER (when it occurs in an intergenic region)\n",
    "- AFF: gene affected\n",
    "- MUT: mutation (in nucleotide if not coding in amino acid if coding)\n",
    "- ANN_TYPE: annotation of the loci, can be intergenic, gene, or cassette (covering the principal elements in the cassette)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data for genome 5\n",
      "Loading data for genome 20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>70</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>3.448276</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-622C&gt;T</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>149</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1.351351</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-543T&gt;C</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1.470588</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-542A&gt;T</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>154</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1.639344</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-538T&gt;C</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>176</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1.754386</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-516_-513delTAATinsCAAC</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508235</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822775</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.588235</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Asn44Ser</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508236</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822787</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>1.156069</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Gly41Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508237</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822830</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>0.609756</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Lys27Thr</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508238</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822867</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Thr15Ala</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508239</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822913</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN684</td>\n",
       "      <td>c.-3810_-3802delATACGCGCGinsTTTCACGCA</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508239 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "1            p2_2_S1     2      70  4.927350e-15   60    58     2  3.448276   \n",
       "2            p2_2_S1     2     149  0.000000e+00   75    74     1  1.351351   \n",
       "3            p2_2_S1     2     150  2.263910e-15   69    68     1  1.470588   \n",
       "4            p2_2_S1     2     154  0.000000e+00   62    61     1  1.639344   \n",
       "5            p2_2_S1     2     176  3.420820e-15   58    57     1  1.754386   \n",
       "...              ...   ...     ...           ...  ...   ...   ...       ...   \n",
       "508235  p3IPTG_A_S1b    18  822775  3.631530e-15  171   170     1  0.588235   \n",
       "508236  p3IPTG_A_S1b    18  822787  1.152910e-15  175   173     2  1.156069   \n",
       "508237  p3IPTG_A_S1b    18  822830  3.857310e-15  165   164     1  0.609756   \n",
       "508238  p3IPTG_A_S1b    18  822867  0.000000e+00  172   171     1  0.584795   \n",
       "508239  p3IPTG_A_S1b    18  822913  0.000000e+00  161   160     1  0.625000   \n",
       "\n",
       "              REF        ALT                    EFF    IMPACT     AFF  \\\n",
       "1               C          T  upstream_gene_variant  MODIFIER  MPN001   \n",
       "2               T          C  upstream_gene_variant  MODIFIER  MPN001   \n",
       "3               A          T  upstream_gene_variant  MODIFIER  MPN001   \n",
       "4               T          C  upstream_gene_variant  MODIFIER  MPN001   \n",
       "5            TAAT       CAAC  upstream_gene_variant  MODIFIER  MPN001   \n",
       "...           ...        ...                    ...       ...     ...   \n",
       "508235      AACGT      GACGC       missense_variant  MODERATE  MPN688   \n",
       "508236          A          G     synonymous_variant       LOW  MPN688   \n",
       "508237          T          G       missense_variant  MODERATE  MPN688   \n",
       "508238      TTTTT      CTTTC       missense_variant  MODERATE  MPN688   \n",
       "508239  CGCGCGTAT  TGCGTGAAA  upstream_gene_variant  MODIFIER  MPN684   \n",
       "\n",
       "                                          MUT    ANN_TYPE  \n",
       "1                                   c.-622C>T  intergenic  \n",
       "2                                   c.-543T>C  intergenic  \n",
       "3                                   c.-542A>T  intergenic  \n",
       "4                                   c.-538T>C  intergenic  \n",
       "5                   c.-516_-513delTAATinsCAAC  intergenic  \n",
       "...                                       ...         ...  \n",
       "508235                             p.Asn44Ser        gene  \n",
       "508236                             p.Gly41Gly        gene  \n",
       "508237                             p.Lys27Thr        gene  \n",
       "508238                             p.Thr15Ala        gene  \n",
       "508239  c.-3810_-3802delATACGCGCGinsTTTCACGCA  intergenic  \n",
       "\n",
       "[508239 rows x 15 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This will concat all the annotated vcf files keeping the fields of interest\n",
    "snpcalls1 = scripts.parse_variations('./filter1/*.raw.vcf.ann', order1, genome=5, fname='./results/data1.pickle')\n",
    "snpcalls2 = scripts.parse_variations('./filter2/*.raw.vcf.ann', order2, genome=20, fname='./results/data2.pickle')\n",
    "snpcalls1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intergenic\n",
      "gene\n",
      "cassette\n"
     ]
    }
   ],
   "source": [
    "for i in list(Counter(snpcalls1['ANN_TYPE']).keys()):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>149</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.449275</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>154</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.612903</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>176</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.724138</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508435</td>\n",
       "      <td>822775</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>GACGC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508436</td>\n",
       "      <td>822787</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.142857</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|synonymous_variant|LOW|soj|MPN688|transcript...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508437</td>\n",
       "      <td>822830</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|missense_variant|MODERATE|soj|MPN688|transcr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508438</td>\n",
       "      <td>822867</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>CTTTC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508439</td>\n",
       "      <td>822913</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508440 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           POS        REF        ALT          QUAL  TOT  REFN  ALTN ALTS  \\\n",
       "0           70          C          T  4.927350e-15   60    58     2    2   \n",
       "1          149          T          C  0.000000e+00   75    74     1    1   \n",
       "2          150          A          T  2.263910e-15   69    68     1    1   \n",
       "3          154          T          C  0.000000e+00   62    61     1    1   \n",
       "4          176       TAAT       CAAC  3.420820e-15   58    57     1    1   \n",
       "...        ...        ...        ...           ...  ...   ...   ...  ...   \n",
       "508435  822775      AACGT      GACGC  3.631530e-15  171   170     1    1   \n",
       "508436  822787          A          G  1.152910e-15  175   173     2    2   \n",
       "508437  822830          T          G  3.857310e-15  165   164     1    1   \n",
       "508438  822867      TTTTT      CTTTC  0.000000e+00  172   171     1    1   \n",
       "508439  822913  CGCGCGTAT  TGCGTGAAA  0.000000e+00  161   160     1    1   \n",
       "\n",
       "            FRAC        SAMPLE   P  \\\n",
       "0       3.333333       p2_2_S1   2   \n",
       "1       1.333333       p2_2_S1   2   \n",
       "2       1.449275       p2_2_S1   2   \n",
       "3       1.612903       p2_2_S1   2   \n",
       "4       1.724138       p2_2_S1   2   \n",
       "...          ...           ...  ..   \n",
       "508435  0.584795  p3IPTG_A_S1b  18   \n",
       "508436  1.142857  p3IPTG_A_S1b  18   \n",
       "508437  0.606061  p3IPTG_A_S1b  18   \n",
       "508438  0.581395  p3IPTG_A_S1b  18   \n",
       "508439  0.621118  p3IPTG_A_S1b  18   \n",
       "\n",
       "                                                      EFF  \n",
       "0       T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "1       C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "2       T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "3       C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "4       CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...  \n",
       "...                                                   ...  \n",
       "508435  GACGC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "508436  G|synonymous_variant|LOW|soj|MPN688|transcript...  \n",
       "508437  G|missense_variant|MODERATE|soj|MPN688|transcr...  \n",
       "508438  CTTTC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "508439  TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...  \n",
       "\n",
       "[508440 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls1\n",
    "# Save supplementary\n",
    "sup1 = './results/supplementary1_variantcalling.xlsx'\n",
    "if not os.path.isfile(sup1):\n",
    "    writer = pd.ExcelWriter(sup1, engine='xlsxwriter')\n",
    "    for c, data in zip(['C5_', 'C20_'], [snpcalls1, snpcalls2]):\n",
    "        for sample in set(data['SAMPLE']):\n",
    "            subdata = data[data['SAMPLE']==sample].copy()\n",
    "            subdata.to_excel(writer, sheet_name=c+sample)\n",
    "    writer.save()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>18805</td>\n",
       "      <td>572048</td>\n",
       "      <td>TTAT</td>\n",
       "      <td>TAAC,CTAG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>112</td>\n",
       "      <td>110</td>\n",
       "      <td>2</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1.785714</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CTAG|upstream_gene_variant|MODIFIER|Gene_56771...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18806</td>\n",
       "      <td>572054</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>109</td>\n",
       "      <td>108</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.917431</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60084</td>\n",
       "      <td>572047</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>72</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.388889</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60085</td>\n",
       "      <td>572049</td>\n",
       "      <td>TAT</td>\n",
       "      <td>AAC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>78</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.282051</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>AAC|upstream_gene_variant|MODIFIER|Gene_567711...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60086</td>\n",
       "      <td>572055</td>\n",
       "      <td>TTCTAAAT</td>\n",
       "      <td>GTCTGAAT,TTCCAAAC,CTCTAAAT</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>79</td>\n",
       "      <td>76</td>\n",
       "      <td>3</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>3.797468</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>CTCTAAAT|upstream_gene_variant|MODIFIER|Gene_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60087</td>\n",
       "      <td>572063</td>\n",
       "      <td>A</td>\n",
       "      <td>G,T</td>\n",
       "      <td>6.238460e-15</td>\n",
       "      <td>80</td>\n",
       "      <td>78</td>\n",
       "      <td>2</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>G|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>117712</td>\n",
       "      <td>572052</td>\n",
       "      <td>TAATT</td>\n",
       "      <td>AAATA</td>\n",
       "      <td>1.808350e-15</td>\n",
       "      <td>261</td>\n",
       "      <td>260</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.383142</td>\n",
       "      <td>p2_2_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>AAATA|upstream_gene_variant|MODIFIER|Gene_5677...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>117713</td>\n",
       "      <td>572058</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>257</td>\n",
       "      <td>256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.389105</td>\n",
       "      <td>p2_2_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>207709</td>\n",
       "      <td>572051</td>\n",
       "      <td>TTA</td>\n",
       "      <td>CTT</td>\n",
       "      <td>1.529410e-14</td>\n",
       "      <td>613</td>\n",
       "      <td>612</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.163132</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>CTT|upstream_gene_variant|MODIFIER|Gene_567711...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>207710</td>\n",
       "      <td>572058</td>\n",
       "      <td>TAAAT</td>\n",
       "      <td>AAAAC</td>\n",
       "      <td>1.225020e-14</td>\n",
       "      <td>619</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161551</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>AAAAC|upstream_gene_variant|MODIFIER|Gene_5677...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>207711</td>\n",
       "      <td>572065</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>583</td>\n",
       "      <td>581</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.343053</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277502</td>\n",
       "      <td>572054</td>\n",
       "      <td>ATTCTA</td>\n",
       "      <td>GTTCAT</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>278</td>\n",
       "      <td>277</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.359712</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>GTTCAT|upstream_gene_variant|MODIFIER|Gene_567...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277503</td>\n",
       "      <td>572063</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>6.632110e-16</td>\n",
       "      <td>272</td>\n",
       "      <td>271</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.367647</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>328006</td>\n",
       "      <td>572046</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>167</td>\n",
       "      <td>166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.598802</td>\n",
       "      <td>p15_F_S1</td>\n",
       "      <td>15</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>377405</td>\n",
       "      <td>572052</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>239</td>\n",
       "      <td>238</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.418410</td>\n",
       "      <td>p15_F_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>377406</td>\n",
       "      <td>572063</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>226</td>\n",
       "      <td>225</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.442478</td>\n",
       "      <td>p15_F_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443874</td>\n",
       "      <td>572049</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4.580790e-15</td>\n",
       "      <td>328</td>\n",
       "      <td>327</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.304878</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443875</td>\n",
       "      <td>572058</td>\n",
       "      <td>TAAATACTA</td>\n",
       "      <td>TAAATCCTC,TAATACTA,TAGATACTA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>305</td>\n",
       "      <td>302</td>\n",
       "      <td>3</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>0.983607</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>TAGATACTA|upstream_gene_variant|MODIFIER|Gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495443</td>\n",
       "      <td>572053</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>6.080020e-15</td>\n",
       "      <td>134</td>\n",
       "      <td>133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.746269</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495444</td>\n",
       "      <td>572063</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>6.434110e-15</td>\n",
       "      <td>129</td>\n",
       "      <td>128</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.775194</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|Gene_567711_5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           POS        REF                           ALT          QUAL  TOT  \\\n",
       "18805   572048       TTAT                     TAAC,CTAG  0.000000e+00  112   \n",
       "18806   572054          A                             G  0.000000e+00  109   \n",
       "60084   572047          T                             A  0.000000e+00   72   \n",
       "60085   572049        TAT                           AAC  0.000000e+00   78   \n",
       "60086   572055   TTCTAAAT    GTCTGAAT,TTCCAAAC,CTCTAAAT  0.000000e+00   79   \n",
       "60087   572063          A                           G,T  6.238460e-15   80   \n",
       "117712  572052      TAATT                         AAATA  1.808350e-15  261   \n",
       "117713  572058          T                             C  0.000000e+00  257   \n",
       "207709  572051        TTA                           CTT  1.529410e-14  613   \n",
       "207710  572058      TAAAT                         AAAAC  1.225020e-14  619   \n",
       "207711  572065          T                             C  0.000000e+00  583   \n",
       "277502  572054     ATTCTA                        GTTCAT  0.000000e+00  278   \n",
       "277503  572063          A                             T  6.632110e-16  272   \n",
       "328006  572046          C                             T  0.000000e+00  167   \n",
       "377405  572052          T                             A  0.000000e+00  239   \n",
       "377406  572063          A                             T  0.000000e+00  226   \n",
       "443874  572049          T                             C  4.580790e-15  328   \n",
       "443875  572058  TAAATACTA  TAAATCCTC,TAATACTA,TAGATACTA  0.000000e+00  305   \n",
       "495443  572053          A                             C  6.080020e-15  134   \n",
       "495444  572063          A                             C  6.434110e-15  129   \n",
       "\n",
       "        REFN  ALTN   ALTS      FRAC        SAMPLE   P  \\\n",
       "18805    110     2    1,1  1.785714       p2_2_S1   2   \n",
       "18806    108     1      1  0.917431       p2_2_S1   2   \n",
       "60084     71     1      1  1.388889       p2_2_S2   2   \n",
       "60085     77     1      1  1.282051       p2_2_S2   2   \n",
       "60086     76     3  1,1,1  3.797468       p2_2_S2   2   \n",
       "60087     78     2    1,1  2.500000       p2_2_S2   2   \n",
       "117712   260     1      1  0.383142       p2_2_S3   2   \n",
       "117713   256     1      1  0.389105       p2_2_S3   2   \n",
       "207709   612     1      1  0.163132       p3_D_S2   3   \n",
       "207710   618     1      1  0.161551       p3_D_S2   3   \n",
       "207711   581     2      2  0.343053       p3_D_S2   3   \n",
       "277502   277     1      1  0.359712       p3_D_S3   3   \n",
       "277503   271     1      1  0.367647       p3_D_S3   3   \n",
       "328006   166     1      1  0.598802      p15_F_S1  15   \n",
       "377405   238     1      1  0.418410      p15_F_S5  15   \n",
       "377406   225     1      1  0.442478      p15_F_S5  15   \n",
       "443874   327     1      1  0.304878   p3IPTG_A_S1  18   \n",
       "443875   302     3  1,1,1  0.983607   p3IPTG_A_S1  18   \n",
       "495443   133     1      1  0.746269  p3IPTG_A_S1b  18   \n",
       "495444   128     1      1  0.775194  p3IPTG_A_S1b  18   \n",
       "\n",
       "                                                      EFF  \n",
       "18805   CTAG|upstream_gene_variant|MODIFIER|Gene_56771...  \n",
       "18806   G|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "60084   A|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "60085   AAC|upstream_gene_variant|MODIFIER|Gene_567711...  \n",
       "60086   CTCTAAAT|upstream_gene_variant|MODIFIER|Gene_5...  \n",
       "60087   G|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "117712  AAATA|upstream_gene_variant|MODIFIER|Gene_5677...  \n",
       "117713  C|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "207709  CTT|upstream_gene_variant|MODIFIER|Gene_567711...  \n",
       "207710  AAAAC|upstream_gene_variant|MODIFIER|Gene_5677...  \n",
       "207711  C|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "277502  GTTCAT|upstream_gene_variant|MODIFIER|Gene_567...  \n",
       "277503  T|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "328006  T|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "377405  A|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "377406  T|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "443874  C|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "443875  TAGATACTA|upstream_gene_variant|MODIFIER|Gene_...  \n",
       "495443  C|upstream_gene_variant|MODIFIER|Gene_567711_5...  \n",
       "495444  C|upstream_gene_variant|MODIFIER|Gene_567711_5...  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example:\n",
    "## Notice the information is grouped by position, we will deal with this in section 3. \n",
    "snpcalls1[snpcalls1['POS'].isin(range(572044, 572066))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>418107</td>\n",
       "      <td>826526</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2.932180e-14</td>\n",
       "      <td>160</td>\n",
       "      <td>152</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>C|stop_lost|HIGH|cysA|MPN685|transcript|MPN685...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>396742</td>\n",
       "      <td>603330</td>\n",
       "      <td>ATTAATAA</td>\n",
       "      <td>GTTAGTAG,ACTAATAA,ATTAAAAA,ATTAATAG,ATTAATAT</td>\n",
       "      <td>5.831640e-15</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>5</td>\n",
       "      <td>1,1,1,1,1</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>GTTAGTAG|upstream_gene_variant|MODIFIER|P02_or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100139</td>\n",
       "      <td>183736</td>\n",
       "      <td>CAACCTCACCCC</td>\n",
       "      <td>CGACGTCAGCCC,GAGGCTCACCCC,CAGCCTCAGCGC,CAACGTC...</td>\n",
       "      <td>1.763150e-15</td>\n",
       "      <td>80</td>\n",
       "      <td>76</td>\n",
       "      <td>4</td>\n",
       "      <td>1,1,1,1</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>GAGGCTCACCCC|missense_variant|MODERATE|P1|MPN1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>394018</td>\n",
       "      <td>572827</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>4.776910e-15</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>G|missense_variant|MODERATE|H08_orf157a|MPN463...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100420</td>\n",
       "      <td>187399</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>5.032580e-15</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>C|missense_variant|MODERATE|orf6|MPN142|transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225401</td>\n",
       "      <td>780968</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>6.689500e+03</td>\n",
       "      <td>214</td>\n",
       "      <td>2</td>\n",
       "      <td>212</td>\n",
       "      <td>209</td>\n",
       "      <td>99.065421</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>413979</td>\n",
       "      <td>780968</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>3.700950e+03</td>\n",
       "      <td>118</td>\n",
       "      <td>1</td>\n",
       "      <td>117</td>\n",
       "      <td>116</td>\n",
       "      <td>99.152542</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163301</td>\n",
       "      <td>780968</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>4.857580e+03</td>\n",
       "      <td>156</td>\n",
       "      <td>1</td>\n",
       "      <td>155</td>\n",
       "      <td>151</td>\n",
       "      <td>99.358974</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163413</td>\n",
       "      <td>786151</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>5.734500e+03</td>\n",
       "      <td>182</td>\n",
       "      <td>1</td>\n",
       "      <td>181</td>\n",
       "      <td>178</td>\n",
       "      <td>99.450549</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>TTT|synonymous_variant|LOW|Gene_785567_786226|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277174</td>\n",
       "      <td>786151</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>TTTGTG</td>\n",
       "      <td>7.229180e+03</td>\n",
       "      <td>228</td>\n",
       "      <td>1</td>\n",
       "      <td>227</td>\n",
       "      <td>224</td>\n",
       "      <td>99.561404</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>TTTGTG|synonymous_variant|LOW|Gene_785567_7862...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>357 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           POS           REF  \\\n",
       "418107  826526             T   \n",
       "396742  603330      ATTAATAA   \n",
       "100139  183736  CAACCTCACCCC   \n",
       "394018  572827             T   \n",
       "100420  187399             A   \n",
       "...        ...           ...   \n",
       "225401  780968         AGTCC   \n",
       "413979  780968         AGTCC   \n",
       "163301  780968         AGTCC   \n",
       "163413  786151           TCT   \n",
       "277174  786151        TCTGTG   \n",
       "\n",
       "                                                      ALT          QUAL  TOT  \\\n",
       "418107                                                  C  2.932180e-14  160   \n",
       "396742       GTTAGTAG,ACTAATAA,ATTAAAAA,ATTAATAG,ATTAATAT  5.831640e-15  100   \n",
       "100139  CGACGTCAGCCC,GAGGCTCACCCC,CAGCCTCAGCGC,CAACGTC...  1.763150e-15   80   \n",
       "394018                                                  G  4.776910e-15   20   \n",
       "100420                                                  C  5.032580e-15  100   \n",
       "...                                                   ...           ...  ...   \n",
       "225401                                              CGTCC  6.689500e+03  214   \n",
       "413979                                              CGTCC  3.700950e+03  118   \n",
       "163301                                              CGTCC  4.857580e+03  156   \n",
       "163413                                                TTT  5.734500e+03  182   \n",
       "277174                                             TTTGTG  7.229180e+03  228   \n",
       "\n",
       "        REFN  ALTN       ALTS       FRAC    SAMPLE   P  \\\n",
       "418107   152     8          8   5.000000  p15_7_S5  15   \n",
       "396742    95     5  1,1,1,1,1   5.000000  p15_7_S5  15   \n",
       "100139    76     4    1,1,1,1   5.000000   p2_6_S4   2   \n",
       "394018    19     1          1   5.000000  p15_7_S5  15   \n",
       "100420    95     5          5   5.000000   p2_6_S4   2   \n",
       "...      ...   ...        ...        ...       ...  ..   \n",
       "225401     2   212        209  99.065421   p3_B_S1   3   \n",
       "413979     1   117        116  99.152542  p15_7_S5  15   \n",
       "163301     1   155        151  99.358974   p2_6_S4   2   \n",
       "163413     1   181        178  99.450549   p2_6_S4   2   \n",
       "277174     1   227        224  99.561404   p3_B_S2   3   \n",
       "\n",
       "                                                      EFF  \n",
       "418107  C|stop_lost|HIGH|cysA|MPN685|transcript|MPN685...  \n",
       "396742  GTTAGTAG|upstream_gene_variant|MODIFIER|P02_or...  \n",
       "100139  GAGGCTCACCCC|missense_variant|MODERATE|P1|MPN1...  \n",
       "394018  G|missense_variant|MODERATE|H08_orf157a|MPN463...  \n",
       "100420  C|missense_variant|MODERATE|orf6|MPN142|transc...  \n",
       "...                                                   ...  \n",
       "225401  CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...  \n",
       "413979  CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...  \n",
       "163301  CGTCC|upstream_gene_variant|MODIFIER|E09_orf30...  \n",
       "163413  TTT|synonymous_variant|LOW|Gene_785567_786226|...  \n",
       "277174  TTTGTG|synonymous_variant|LOW|Gene_785567_7862...  \n",
       "\n",
       "[357 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls2[(snpcalls2['FRAC']>=5) & (snpcalls2['FRAC']<100)].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Effect of the mutations\n",
    "\n",
    "Deconvolute alternative and total read count to evaluate the population fraction and impact of each variant within the cassette."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [SeqFeature(FeatureLocation(ExactPosition(565549), ExactPosition(566678), strand=-1), type='CDS'),\n",
       "  'CDS'],\n",
       " 1: [SeqFeature(FeatureLocation(ExactPosition(566678), ExactPosition(566721), strand=-1), type='promoter'),\n",
       "  'promoter'],\n",
       " 2: [SeqFeature(FeatureLocation(ExactPosition(567711), ExactPosition(571818), strand=-1), type='CDS'),\n",
       "  'CDS'],\n",
       " 3: [SeqFeature(FeatureLocation(ExactPosition(571818), ExactPosition(571934), strand=-1), type='promoter'),\n",
       "  'promoter'],\n",
       " 4: [SeqFeature(FeatureLocation(ExactPosition(571942), ExactPosition(572044), strand=-1), type='misc_RNA'),\n",
       "  'misc_RNA'],\n",
       " 5: [SeqFeature(FeatureLocation(ExactPosition(572044), ExactPosition(572066), strand=-1), type='promoter'),\n",
       "  'promoter']}"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "correspondance(20, positions=cas1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-114-acf61a0d8854>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-114-acf61a0d8854>\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    ann[]\u001b[0m\n\u001b[0m        ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for k, v in correspondance(20, positions=set(range(1, 1000000))):\n",
    "    if int(v.location.start) in cas1:\n",
    "        ann[]\n",
    "    else:\n",
    "        ann = {v.qualifiers['locus_tag'][0]:[int(v.location.start), int(v.location.end), '-'] for k, v in ann.items() if 'locus_tag' in v.qualifiers}\n",
    "    if positions:\n",
    "        annotations = correspondance(genome=genome, positions=positions)\n",
    "        annotations = {}\n",
    "    corresponde = {'Gene_{}_{}'.format(v[0], v[1]-1):k for k, v in annotations.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GAAAAAAAAACTAG': ['frameshift_variant',\n",
       "  'HIGH',\n",
       "  'transcript',\n",
       "  'p.Leu1675fs',\n",
       "  'gene']}"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_effect(snpcalls1.iloc[102962].EFF, 370436)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>70</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>3.448276</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-622C&gt;T</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>149</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1.351351</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-543T&gt;C</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1.470588</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-542A&gt;T</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>154</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1.639344</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-538T&gt;C</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>176</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1.754386</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-516_-513delTAATinsCAAC</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508235</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822775</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.588235</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Asn44Ser</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508236</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822787</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>1.156069</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Gly41Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508237</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822830</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>0.609756</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Lys27Thr</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508238</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822867</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Thr15Ala</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508239</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>822913</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN684</td>\n",
       "      <td>c.-3810_-3802delATACGCGCGinsTTTCACGCA</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508239 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "1            p2_2_S1     2      70  4.927350e-15   60    58     2  3.448276   \n",
       "2            p2_2_S1     2     149  0.000000e+00   75    74     1  1.351351   \n",
       "3            p2_2_S1     2     150  2.263910e-15   69    68     1  1.470588   \n",
       "4            p2_2_S1     2     154  0.000000e+00   62    61     1  1.639344   \n",
       "5            p2_2_S1     2     176  3.420820e-15   58    57     1  1.754386   \n",
       "...              ...   ...     ...           ...  ...   ...   ...       ...   \n",
       "508235  p3IPTG_A_S1b    18  822775  3.631530e-15  171   170     1  0.588235   \n",
       "508236  p3IPTG_A_S1b    18  822787  1.152910e-15  175   173     2  1.156069   \n",
       "508237  p3IPTG_A_S1b    18  822830  3.857310e-15  165   164     1  0.609756   \n",
       "508238  p3IPTG_A_S1b    18  822867  0.000000e+00  172   171     1  0.584795   \n",
       "508239  p3IPTG_A_S1b    18  822913  0.000000e+00  161   160     1  0.625000   \n",
       "\n",
       "              REF        ALT                    EFF    IMPACT     AFF  \\\n",
       "1               C          T  upstream_gene_variant  MODIFIER  MPN001   \n",
       "2               T          C  upstream_gene_variant  MODIFIER  MPN001   \n",
       "3               A          T  upstream_gene_variant  MODIFIER  MPN001   \n",
       "4               T          C  upstream_gene_variant  MODIFIER  MPN001   \n",
       "5            TAAT       CAAC  upstream_gene_variant  MODIFIER  MPN001   \n",
       "...           ...        ...                    ...       ...     ...   \n",
       "508235      AACGT      GACGC       missense_variant  MODERATE  MPN688   \n",
       "508236          A          G     synonymous_variant       LOW  MPN688   \n",
       "508237          T          G       missense_variant  MODERATE  MPN688   \n",
       "508238      TTTTT      CTTTC       missense_variant  MODERATE  MPN688   \n",
       "508239  CGCGCGTAT  TGCGTGAAA  upstream_gene_variant  MODIFIER  MPN684   \n",
       "\n",
       "                                          MUT    ANN_TYPE  \n",
       "1                                   c.-622C>T  intergenic  \n",
       "2                                   c.-543T>C  intergenic  \n",
       "3                                   c.-542A>T  intergenic  \n",
       "4                                   c.-538T>C  intergenic  \n",
       "5                   c.-516_-513delTAATinsCAAC  intergenic  \n",
       "...                                       ...         ...  \n",
       "508235                             p.Asn44Ser        gene  \n",
       "508236                             p.Gly41Gly        gene  \n",
       "508237                             p.Lys27Thr        gene  \n",
       "508238                             p.Thr15Ala        gene  \n",
       "508239  c.-3810_-3802delATACGCGCGinsTTTCACGCA  intergenic  \n",
       "\n",
       "[508239 rows x 15 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simplify_effect(snpcalls1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "effect1 = simplify_effect(snpcalls1.head(200))\n",
    "effect2 = simplify_effect(snpcalls2, genome=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "effect2.to_csv('./results/effect2.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>39</td>\n",
       "      <td>2.676060e-14</td>\n",
       "      <td>228</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.884956</td>\n",
       "      <td>TGCT</td>\n",
       "      <td>AGCA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-653_-650delTGCTinsAGCA</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>47</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>248</td>\n",
       "      <td>246</td>\n",
       "      <td>2</td>\n",
       "      <td>0.813008</td>\n",
       "      <td>ACGTG</td>\n",
       "      <td>TCGTA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-645_-641delACGTGinsTCGTA</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>292</td>\n",
       "      <td>290</td>\n",
       "      <td>2</td>\n",
       "      <td>0.689655</td>\n",
       "      <td>CGCGAAAAAGA</td>\n",
       "      <td>TCCATAGTACT</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-621_-611delCGCGAAAAAGAinsTCCATAGTACT</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>115</td>\n",
       "      <td>7.127220e-15</td>\n",
       "      <td>323</td>\n",
       "      <td>321</td>\n",
       "      <td>2</td>\n",
       "      <td>0.623053</td>\n",
       "      <td>TTTGCCG</td>\n",
       "      <td>TCAGCCC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-576_-571delTTGCCGinsCAGCCC</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>133</td>\n",
       "      <td>1.401310e-14</td>\n",
       "      <td>343</td>\n",
       "      <td>341</td>\n",
       "      <td>2</td>\n",
       "      <td>0.586510</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN001</td>\n",
       "      <td>c.-559A&gt;T</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>418219</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>829438</td>\n",
       "      <td>1.974090e-15</td>\n",
       "      <td>150</td>\n",
       "      <td>148</td>\n",
       "      <td>2</td>\n",
       "      <td>1.351351</td>\n",
       "      <td>ACAAGC</td>\n",
       "      <td>AGAAGT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Val26Leu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>418221</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>829469</td>\n",
       "      <td>1.838480e-15</td>\n",
       "      <td>170</td>\n",
       "      <td>167</td>\n",
       "      <td>3</td>\n",
       "      <td>1.796407</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Thr16Ala</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>418222</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>829474</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>162</td>\n",
       "      <td>160</td>\n",
       "      <td>2</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Lys14Arg</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>418224</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>829501</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>160</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>2.564103</td>\n",
       "      <td>AAAGAAATAAT</td>\n",
       "      <td>GAGGAAATAAT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN688</td>\n",
       "      <td>p.Phe5Ser</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>418225</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>829522</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>163</td>\n",
       "      <td>161</td>\n",
       "      <td>2</td>\n",
       "      <td>1.242236</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN684</td>\n",
       "      <td>c.-3806G&gt;A</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103682 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "4        p2_6_S2     2      39  2.676060e-14  228   226     2  0.884956   \n",
       "5        p2_6_S2     2      47  0.000000e+00  248   246     2  0.813008   \n",
       "8        p2_6_S2     2      71  0.000000e+00  292   290     2  0.689655   \n",
       "11       p2_6_S2     2     115  7.127220e-15  323   321     2  0.623053   \n",
       "12       p2_6_S2     2     133  1.401310e-14  343   341     2  0.586510   \n",
       "...          ...   ...     ...           ...  ...   ...   ...       ...   \n",
       "418219  p15_7_S5    15  829438  1.974090e-15  150   148     2  1.351351   \n",
       "418221  p15_7_S5    15  829469  1.838480e-15  170   167     3  1.796407   \n",
       "418222  p15_7_S5    15  829474  0.000000e+00  162   160     2  1.250000   \n",
       "418224  p15_7_S5    15  829501  0.000000e+00  160   156     4  2.564103   \n",
       "418225  p15_7_S5    15  829522  0.000000e+00  163   161     2  1.242236   \n",
       "\n",
       "                REF          ALT                    EFF    IMPACT     AFF  \\\n",
       "4              TGCT         AGCA  upstream_gene_variant  MODIFIER  MPN001   \n",
       "5             ACGTG        TCGTA  upstream_gene_variant  MODIFIER  MPN001   \n",
       "8       CGCGAAAAAGA  TCCATAGTACT  upstream_gene_variant  MODIFIER  MPN001   \n",
       "11          TTTGCCG      TCAGCCC  upstream_gene_variant  MODIFIER  MPN001   \n",
       "12                A            T  upstream_gene_variant  MODIFIER  MPN001   \n",
       "...             ...          ...                    ...       ...     ...   \n",
       "418219       ACAAGC       AGAAGT       missense_variant  MODERATE  MPN688   \n",
       "418221            T            C       missense_variant  MODERATE  MPN688   \n",
       "418222            T            C       missense_variant  MODERATE  MPN688   \n",
       "418224  AAAGAAATAAT  GAGGAAATAAT       missense_variant  MODERATE  MPN688   \n",
       "418225            C            T  upstream_gene_variant  MODIFIER  MPN684   \n",
       "\n",
       "                                            MUT    ANN_TYPE  \n",
       "4                     c.-653_-650delTGCTinsAGCA  intergenic  \n",
       "5                   c.-645_-641delACGTGinsTCGTA  intergenic  \n",
       "8       c.-621_-611delCGCGAAAAAGAinsTCCATAGTACT  intergenic  \n",
       "11                c.-576_-571delTTGCCGinsCAGCCC  intergenic  \n",
       "12                                    c.-559A>T  intergenic  \n",
       "...                                         ...         ...  \n",
       "418219                               p.Val26Leu        gene  \n",
       "418221                               p.Thr16Ala        gene  \n",
       "418222                               p.Lys14Arg        gene  \n",
       "418224                                p.Phe5Ser        gene  \n",
       "418225                               c.-3806G>A  intergenic  \n",
       "\n",
       "[103682 rows x 15 columns]"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect2[effect2['ALTN']>=2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1533</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566618</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>985</td>\n",
       "      <td>984</td>\n",
       "      <td>1</td>\n",
       "      <td>0.101626</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Val21Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1491</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566174</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>962</td>\n",
       "      <td>961</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104058</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Asn169Tyr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1532</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566617</td>\n",
       "      <td>1.055200e-15</td>\n",
       "      <td>960</td>\n",
       "      <td>959</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104275</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Val21Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1469</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566002</td>\n",
       "      <td>1.296050e-15</td>\n",
       "      <td>960</td>\n",
       "      <td>959</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104275</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ala226Gly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1529</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566608</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>958</td>\n",
       "      <td>957</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104493</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Tyr24Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4176</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566416</td>\n",
       "      <td>3.471600e-14</td>\n",
       "      <td>281</td>\n",
       "      <td>255</td>\n",
       "      <td>26</td>\n",
       "      <td>10.196078</td>\n",
       "      <td>GGTGCGT</td>\n",
       "      <td>CGTTCTT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.HisAlaPro86GlnGluArg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4289</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>569127</td>\n",
       "      <td>8.076710e-15</td>\n",
       "      <td>235</td>\n",
       "      <td>210</td>\n",
       "      <td>25</td>\n",
       "      <td>11.904762</td>\n",
       "      <td>CAAACT</td>\n",
       "      <td>CAACT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Phe897fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3793</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>463</td>\n",
       "      <td>71</td>\n",
       "      <td>15.334773</td>\n",
       "      <td>ATCAAACT</td>\n",
       "      <td>GTCAACC</td>\n",
       "      <td>frameshift_variant&amp;missense_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Lys896fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4146</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>271</td>\n",
       "      <td>224</td>\n",
       "      <td>47</td>\n",
       "      <td>20.982143</td>\n",
       "      <td>CGCAA</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208MetCys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3549</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>1.947940e-13</td>\n",
       "      <td>559</td>\n",
       "      <td>446</td>\n",
       "      <td>113</td>\n",
       "      <td>25.336323</td>\n",
       "      <td>CGCAAA</td>\n",
       "      <td>TGCTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4462 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "1533       p3_D_S2     3  566618  0.000000e+00  985   984     1   0.101626   \n",
       "1491       p3_D_S2     3  566174  0.000000e+00  962   961     1   0.104058   \n",
       "1532       p3_D_S2     3  566617  1.055200e-15  960   959     1   0.104275   \n",
       "1469       p3_D_S2     3  566002  1.296050e-15  960   959     1   0.104275   \n",
       "1529       p3_D_S2     3  566608  0.000000e+00  958   957     1   0.104493   \n",
       "...            ...   ...     ...           ...  ...   ...   ...        ...   \n",
       "4176  p3IPTG_A_S1b    18  566416  3.471600e-14  281   255    26  10.196078   \n",
       "4289  p3IPTG_A_S1b    18  569127  8.076710e-15  235   210    25  11.904762   \n",
       "3793   p3IPTG_A_S1    18  569125  0.000000e+00  534   463    71  15.334773   \n",
       "4146  p3IPTG_A_S1b    18  566053  0.000000e+00  271   224    47  20.982143   \n",
       "3549   p3IPTG_A_S1    18  566053  1.947940e-13  559   446   113  25.336323   \n",
       "\n",
       "           REF      ALT                                  EFF    IMPACT    AFF  \\\n",
       "1533         C        T                     missense_variant  MODERATE  LacI4   \n",
       "1491         T        A                     missense_variant  MODERATE  LacI4   \n",
       "1532         A        G                     missense_variant  MODERATE  LacI4   \n",
       "1469         G        C                     missense_variant  MODERATE  LacI4   \n",
       "1529         T        A                     missense_variant  MODERATE  LacI4   \n",
       "...        ...      ...                                  ...       ...    ...   \n",
       "4176   GGTGCGT  CGTTCTT                     missense_variant  MODERATE  LacI4   \n",
       "4289    CAAACT    CAACT                   frameshift_variant      HIGH  cas9B   \n",
       "3793  ATCAAACT  GTCAACC  frameshift_variant&missense_variant      HIGH  cas9B   \n",
       "4146     CGCAA    CACAT                     missense_variant  MODERATE  LacI4   \n",
       "3549    CGCAAA   TGCTAT                          stop_gained      HIGH  LacI4   \n",
       "\n",
       "                         MUT  \n",
       "1533              p.Val21Met  \n",
       "1491             p.Asn169Tyr  \n",
       "1532              p.Val21Ala  \n",
       "1469             p.Ala226Gly  \n",
       "1529              p.Tyr24Phe  \n",
       "...                      ...  \n",
       "4176  p.HisAlaPro86GlnGluArg  \n",
       "4289              p.Phe897fs  \n",
       "3793              p.Lys896fs  \n",
       "4146       p.LeuArg208MetCys  \n",
       "3549            p.LeuArg208*  \n",
       "\n",
       "[4462 rows x 14 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# High impact mutations in Cas Par in samples C5\n",
    "effect1.sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "41542d7e65264784b0e4649fb3f6c70c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f1bd89350f0>"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure(figsize=(10,10))\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='PASS', data=effect1[(effect1['IMPACT']=='HIGH') & (effect1['FRAC']>=1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>40</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566670</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>226</td>\n",
       "      <td>225</td>\n",
       "      <td>1</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>CTTTGC</td>\n",
       "      <td>CTTGC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Lys3fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>565557</td>\n",
       "      <td>2.712640e-15</td>\n",
       "      <td>155</td>\n",
       "      <td>154</td>\n",
       "      <td>1</td>\n",
       "      <td>0.649351</td>\n",
       "      <td>TCGCT</td>\n",
       "      <td>CCGCC</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter373Trpext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>566676</td>\n",
       "      <td>1.994980e-14</td>\n",
       "      <td>226</td>\n",
       "      <td>224</td>\n",
       "      <td>2</td>\n",
       "      <td>0.892857</td>\n",
       "      <td>CATA</td>\n",
       "      <td>TATT</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Met1?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>566634</td>\n",
       "      <td>2.558530e-14</td>\n",
       "      <td>112</td>\n",
       "      <td>111</td>\n",
       "      <td>1</td>\n",
       "      <td>0.900901</td>\n",
       "      <td>AGGCT</td>\n",
       "      <td>AGGGCT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Val16fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>58</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>567734</td>\n",
       "      <td>5.514750e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>78</td>\n",
       "      <td>1</td>\n",
       "      <td>1.282051</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Leu1362*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>565585</td>\n",
       "      <td>1.952790e-14</td>\n",
       "      <td>72</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>1.408451</td>\n",
       "      <td>ACTTG</td>\n",
       "      <td>GCTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.GlnVal364*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>52</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>565556</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>59</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>1.724138</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter375Lysext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>565556</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>97</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "      <td>2.105263</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter375Lysext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>567751</td>\n",
       "      <td>2.863080e-15</td>\n",
       "      <td>100</td>\n",
       "      <td>97</td>\n",
       "      <td>3</td>\n",
       "      <td>3.092784</td>\n",
       "      <td>ATAAAG</td>\n",
       "      <td>TTAATA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.LeuTyr1355*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC     REF  \\\n",
       "40   p3_B_S2     3  566670  0.000000e+00  226   225     1  0.444444  CTTTGC   \n",
       "1    p2_6_S2     2  565557  2.712640e-15  155   154     1  0.649351   TCGCT   \n",
       "29   p3_B_S1     3  566676  1.994980e-14  226   224     2  0.892857    CATA   \n",
       "27   p3_B_S1     3  566634  2.558530e-14  112   111     1  0.900901   AGGCT   \n",
       "58  p15_7_S5    15  567734  5.514750e-15   79    78     1  1.282051       A   \n",
       "10   p2_6_S4     2  565585  1.952790e-14   72    71     1  1.408451   ACTTG   \n",
       "52  p15_7_S5    15  565556  0.000000e+00   59    58     1  1.724138       A   \n",
       "7    p2_6_S4     2  565556  0.000000e+00   97    95     2  2.105263       A   \n",
       "24   p2_6_S4     2  567751  2.863080e-15  100    97     3  3.092784  ATAAAG   \n",
       "\n",
       "       ALT                 EFF IMPACT    AFF               MUT  \n",
       "40   CTTGC  frameshift_variant   HIGH  LacI4          p.Lys3fs  \n",
       "1    CCGCC           stop_lost   HIGH  LacI4  p.Ter373Trpext*?  \n",
       "29    TATT          start_lost   HIGH  LacI4           p.Met1?  \n",
       "27  AGGGCT  frameshift_variant   HIGH  LacI4         p.Val16fs  \n",
       "58       T         stop_gained   HIGH  cas9B        p.Leu1362*  \n",
       "10   GCTTA         stop_gained   HIGH  LacI4      p.GlnVal364*  \n",
       "52       T           stop_lost   HIGH  LacI4  p.Ter375Lysext*?  \n",
       "7        T           stop_lost   HIGH  LacI4  p.Ter375Lysext*?  \n",
       "24  TTAATA         stop_gained   HIGH  cas9B     p.LeuTyr1355*  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect21[effect21['IMPACT']=='HIGH'].sort_values('FRAC')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>118</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>786095</td>\n",
       "      <td>3.616460e-15</td>\n",
       "      <td>404</td>\n",
       "      <td>403</td>\n",
       "      <td>1</td>\n",
       "      <td>0.248139</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Tyr176*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>130</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>786200</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>401</td>\n",
       "      <td>400</td>\n",
       "      <td>1</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Tyr211*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>785572</td>\n",
       "      <td>7.669890e-15</td>\n",
       "      <td>336</td>\n",
       "      <td>335</td>\n",
       "      <td>1</td>\n",
       "      <td>0.298507</td>\n",
       "      <td>AGAAAAA</td>\n",
       "      <td>TGTAAT</td>\n",
       "      <td>frameshift_variant&amp;stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Glu2fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>785610</td>\n",
       "      <td>1.567030e-15</td>\n",
       "      <td>322</td>\n",
       "      <td>321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.311526</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln15*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>55</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>785614</td>\n",
       "      <td>1.567030e-15</td>\n",
       "      <td>322</td>\n",
       "      <td>321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.311526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Trp16*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>59</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>785651</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>315</td>\n",
       "      <td>314</td>\n",
       "      <td>1</td>\n",
       "      <td>0.318471</td>\n",
       "      <td>TGCTC</td>\n",
       "      <td>AGCTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln30*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>93</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>785881</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>293</td>\n",
       "      <td>292</td>\n",
       "      <td>1</td>\n",
       "      <td>0.342466</td>\n",
       "      <td>TCTG</td>\n",
       "      <td>ACTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.LeuTrp105*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>384</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>786189</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>280</td>\n",
       "      <td>279</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358423</td>\n",
       "      <td>TTAC</td>\n",
       "      <td>CTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln209*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>331</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785573</td>\n",
       "      <td>2.991160e-15</td>\n",
       "      <td>253</td>\n",
       "      <td>252</td>\n",
       "      <td>1</td>\n",
       "      <td>0.396825</td>\n",
       "      <td>GAAAAAAATCAC</td>\n",
       "      <td>GAAAAAAAATCAC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Ile5fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>337</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>252</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0.398406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln26*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>519</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785945</td>\n",
       "      <td>3.239540e-15</td>\n",
       "      <td>250</td>\n",
       "      <td>249</td>\n",
       "      <td>1</td>\n",
       "      <td>0.401606</td>\n",
       "      <td>TTACG</td>\n",
       "      <td>ATACA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.CysTyrGly126*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>506</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785676</td>\n",
       "      <td>2.644840e-15</td>\n",
       "      <td>232</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>0.432900</td>\n",
       "      <td>GTTC</td>\n",
       "      <td>ATTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.ValGln37*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>520</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785969</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>223</td>\n",
       "      <td>222</td>\n",
       "      <td>1</td>\n",
       "      <td>0.450450</td>\n",
       "      <td>CCCTAAA</td>\n",
       "      <td>TCCTTAG</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Lys136*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>523</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>786017</td>\n",
       "      <td>9.833970e-16</td>\n",
       "      <td>219</td>\n",
       "      <td>218</td>\n",
       "      <td>1</td>\n",
       "      <td>0.458716</td>\n",
       "      <td>GGTG</td>\n",
       "      <td>AGTC</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Trp150*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>502</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>217</td>\n",
       "      <td>216</td>\n",
       "      <td>1</td>\n",
       "      <td>0.462963</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln26*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>129</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>786195</td>\n",
       "      <td>9.643510e-15</td>\n",
       "      <td>405</td>\n",
       "      <td>403</td>\n",
       "      <td>2</td>\n",
       "      <td>0.496278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln210*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>191</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>785570</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>202</td>\n",
       "      <td>201</td>\n",
       "      <td>1</td>\n",
       "      <td>0.497512</td>\n",
       "      <td>GGAGA</td>\n",
       "      <td>AGAGT</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.MetGluLys1?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>260</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>786095</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>199</td>\n",
       "      <td>198</td>\n",
       "      <td>1</td>\n",
       "      <td>0.505051</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Tyr176*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>197</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>785616</td>\n",
       "      <td>1.132430e-16</td>\n",
       "      <td>191</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>CAT</td>\n",
       "      <td>TAA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.His17*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>213</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>785701</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>188</td>\n",
       "      <td>187</td>\n",
       "      <td>1</td>\n",
       "      <td>0.534759</td>\n",
       "      <td>TAA</td>\n",
       "      <td>AAG</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.LeuLys45*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>785643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>185</td>\n",
       "      <td>184</td>\n",
       "      <td>1</td>\n",
       "      <td>0.543478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln26*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>138</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>786293</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179</td>\n",
       "      <td>178</td>\n",
       "      <td>1</td>\n",
       "      <td>0.561798</td>\n",
       "      <td>TATTT</td>\n",
       "      <td>AATAA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.ValPhe6*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>601</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>785727</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>162</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Lys54*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>139</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>786316</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>152</td>\n",
       "      <td>151</td>\n",
       "      <td>1</td>\n",
       "      <td>0.662252</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.Lys14*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>595</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>785643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>152</td>\n",
       "      <td>151</td>\n",
       "      <td>1</td>\n",
       "      <td>0.662252</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln26*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>393</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>787394</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>150</td>\n",
       "      <td>149</td>\n",
       "      <td>1</td>\n",
       "      <td>0.671141</td>\n",
       "      <td>AGCG</td>\n",
       "      <td>TGCA</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.TerArg373LeuGlnext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>437</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>785709</td>\n",
       "      <td>1.702880e-15</td>\n",
       "      <td>258</td>\n",
       "      <td>256</td>\n",
       "      <td>2</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GA</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Val48fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>438</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>785714</td>\n",
       "      <td>1.642610e-15</td>\n",
       "      <td>255</td>\n",
       "      <td>253</td>\n",
       "      <td>2</td>\n",
       "      <td>0.790514</td>\n",
       "      <td>GAAAAATAAGC</td>\n",
       "      <td>GAAAATAAGC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Asn51fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>627</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>786046</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>116</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>TGGCC</td>\n",
       "      <td>TGCC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Ala161fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>294</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>787393</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>115</td>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>0.877193</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.Ter373Glnext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>504</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785655</td>\n",
       "      <td>5.703090e-15</td>\n",
       "      <td>223</td>\n",
       "      <td>221</td>\n",
       "      <td>2</td>\n",
       "      <td>0.904977</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Gln30*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>388</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>786318</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>98</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>1.030928</td>\n",
       "      <td>GCCTG</td>\n",
       "      <td>GCCCTG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.Val16fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>525</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>786040</td>\n",
       "      <td>3.744300e-15</td>\n",
       "      <td>195</td>\n",
       "      <td>193</td>\n",
       "      <td>2</td>\n",
       "      <td>1.036269</td>\n",
       "      <td>TAAAC</td>\n",
       "      <td>AAAAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Leu158*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>245</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>785969</td>\n",
       "      <td>1.536900e-15</td>\n",
       "      <td>195</td>\n",
       "      <td>193</td>\n",
       "      <td>2</td>\n",
       "      <td>1.036269</td>\n",
       "      <td>CCCTA</td>\n",
       "      <td>TCCTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Lys136*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>143</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>787387</td>\n",
       "      <td>1.217490e-14</td>\n",
       "      <td>191</td>\n",
       "      <td>189</td>\n",
       "      <td>2</td>\n",
       "      <td>1.058201</td>\n",
       "      <td>GGTCAATA</td>\n",
       "      <td>AGACAACA</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.GlyGlnTer371ArgGlnGlnext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>786211</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>190</td>\n",
       "      <td>188</td>\n",
       "      <td>2</td>\n",
       "      <td>1.063830</td>\n",
       "      <td>GGCA</td>\n",
       "      <td>AGCT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.TrpGln215*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>347</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785712</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>264</td>\n",
       "      <td>261</td>\n",
       "      <td>3</td>\n",
       "      <td>1.149425</td>\n",
       "      <td>AAGAAAA</td>\n",
       "      <td>TAGTAAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.LysLysAsn49*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>528</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>786075</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>230</td>\n",
       "      <td>227</td>\n",
       "      <td>3</td>\n",
       "      <td>1.321586</td>\n",
       "      <td>GTTTTCA</td>\n",
       "      <td>GTTTCA</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.Phe171fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>654</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>787399</td>\n",
       "      <td>1.898760e-15</td>\n",
       "      <td>74</td>\n",
       "      <td>73</td>\n",
       "      <td>1</td>\n",
       "      <td>1.369863</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.Ter375Lysext*?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>292</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>787364</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>73</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>1.388889</td>\n",
       "      <td>GTC</td>\n",
       "      <td>ATT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.ArgGln363*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>284</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>786329</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>57</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>1.785714</td>\n",
       "      <td>TATA</td>\n",
       "      <td>AATT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>lacI4</td>\n",
       "      <td>p.LeuTyr18*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>619</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>785966</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>126</td>\n",
       "      <td>123</td>\n",
       "      <td>3</td>\n",
       "      <td>2.439024</td>\n",
       "      <td>TTTCCCTAAA</td>\n",
       "      <td>CTTCTCTTAA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.ProLys135*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508</td>\n",
       "      <td>p15_7_S3</td>\n",
       "      <td>15</td>\n",
       "      <td>785697</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>258</td>\n",
       "      <td>250</td>\n",
       "      <td>8</td>\n",
       "      <td>3.200000</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>ATTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cat</td>\n",
       "      <td>p.PheLeu44*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "118   p2_6_S2     2  786095  3.616460e-15  404   403     1  0.248139   \n",
       "130   p2_6_S2     2  786200  0.000000e+00  401   400     1  0.250000   \n",
       "47    p2_6_S2     2  785572  7.669890e-15  336   335     1  0.298507   \n",
       "54    p2_6_S2     2  785610  1.567030e-15  322   321     1  0.311526   \n",
       "55    p2_6_S2     2  785614  1.567030e-15  322   321     1  0.311526   \n",
       "59    p2_6_S2     2  785651  0.000000e+00  315   314     1  0.318471   \n",
       "93    p2_6_S2     2  785881  0.000000e+00  293   292     1  0.342466   \n",
       "384   p3_B_S1     3  786189  0.000000e+00  280   279     1  0.358423   \n",
       "331   p3_B_S1     3  785573  2.991160e-15  253   252     1  0.396825   \n",
       "337   p3_B_S1     3  785643  0.000000e+00  252   251     1  0.398406   \n",
       "519  p15_7_S3    15  785945  3.239540e-15  250   249     1  0.401606   \n",
       "506  p15_7_S3    15  785676  2.644840e-15  232   231     1  0.432900   \n",
       "520  p15_7_S3    15  785969  0.000000e+00  223   222     1  0.450450   \n",
       "523  p15_7_S3    15  786017  9.833970e-16  219   218     1  0.458716   \n",
       "502  p15_7_S3    15  785643  0.000000e+00  217   216     1  0.462963   \n",
       "129   p2_6_S2     2  786195  9.643510e-15  405   403     2  0.496278   \n",
       "191   p2_6_S4     2  785570  0.000000e+00  202   201     1  0.497512   \n",
       "260   p2_6_S4     2  786095  0.000000e+00  199   198     1  0.505051   \n",
       "197   p2_6_S4     2  785616  1.132430e-16  191   190     1  0.526316   \n",
       "213   p2_6_S4     2  785701  0.000000e+00  188   187     1  0.534759   \n",
       "200   p2_6_S4     2  785643  0.000000e+00  185   184     1  0.543478   \n",
       "138   p2_6_S2     2  786293  0.000000e+00  179   178     1  0.561798   \n",
       "601  p15_7_S5    15  785727  0.000000e+00  162   161     1  0.621118   \n",
       "139   p2_6_S2     2  786316  0.000000e+00  152   151     1  0.662252   \n",
       "595  p15_7_S5    15  785643  0.000000e+00  152   151     1  0.662252   \n",
       "393   p3_B_S1     3  787394  0.000000e+00  150   149     1  0.671141   \n",
       "437   p3_B_S2     3  785709  1.702880e-15  258   256     2  0.781250   \n",
       "438   p3_B_S2     3  785714  1.642610e-15  255   253     2  0.790514   \n",
       "627  p15_7_S5    15  786046  0.000000e+00  116   115     1  0.869565   \n",
       "294   p2_6_S4     2  787393  0.000000e+00  115   114     1  0.877193   \n",
       "504  p15_7_S3    15  785655  5.703090e-15  223   221     2  0.904977   \n",
       "388   p3_B_S1     3  786318  0.000000e+00   98    97     1  1.030928   \n",
       "525  p15_7_S3    15  786040  3.744300e-15  195   193     2  1.036269   \n",
       "245   p2_6_S4     2  785969  1.536900e-15  195   193     2  1.036269   \n",
       "143   p2_6_S2     2  787387  1.217490e-14  191   189     2  1.058201   \n",
       "277   p2_6_S4     2  786211  0.000000e+00  190   188     2  1.063830   \n",
       "347   p3_B_S1     3  785712  0.000000e+00  264   261     3  1.149425   \n",
       "528  p15_7_S3    15  786075  0.000000e+00  230   227     3  1.321586   \n",
       "654  p15_7_S5    15  787399  1.898760e-15   74    73     1  1.369863   \n",
       "292   p2_6_S4     2  787364  0.000000e+00   73    72     1  1.388889   \n",
       "284   p2_6_S4     2  786329  0.000000e+00   57    56     1  1.785714   \n",
       "619  p15_7_S5    15  785966  0.000000e+00  126   123     3  2.439024   \n",
       "508  p15_7_S3    15  785697  0.000000e+00  258   250     8  3.200000   \n",
       "\n",
       "              REF            ALT                             EFF IMPACT  \\\n",
       "118             T              G                     stop_gained   HIGH   \n",
       "130             C              G                     stop_gained   HIGH   \n",
       "47        AGAAAAA         TGTAAT  frameshift_variant&stop_gained   HIGH   \n",
       "54              C              T                     stop_gained   HIGH   \n",
       "55              G              A                     stop_gained   HIGH   \n",
       "59          TGCTC          AGCTT                     stop_gained   HIGH   \n",
       "93           TCTG           ACTA                     stop_gained   HIGH   \n",
       "384          TTAC           CTAT                     stop_gained   HIGH   \n",
       "331  GAAAAAAATCAC  GAAAAAAAATCAC              frameshift_variant   HIGH   \n",
       "337             C              T                     stop_gained   HIGH   \n",
       "519         TTACG          ATACA                     stop_gained   HIGH   \n",
       "506          GTTC           ATTT                     stop_gained   HIGH   \n",
       "520       CCCTAAA        TCCTTAG                     stop_gained   HIGH   \n",
       "523          GGTG           AGTC                     stop_gained   HIGH   \n",
       "502             C              T                     stop_gained   HIGH   \n",
       "129             C              T                     stop_gained   HIGH   \n",
       "191         GGAGA          AGAGT                      start_lost   HIGH   \n",
       "260             T              A                     stop_gained   HIGH   \n",
       "197           CAT            TAA                     stop_gained   HIGH   \n",
       "213           TAA            AAG                     stop_gained   HIGH   \n",
       "200             C              T                     stop_gained   HIGH   \n",
       "138         TATTT          AATAA                     stop_gained   HIGH   \n",
       "601             A              T                     stop_gained   HIGH   \n",
       "139             A              T                     stop_gained   HIGH   \n",
       "595             C              T                     stop_gained   HIGH   \n",
       "393          AGCG           TGCA                       stop_lost   HIGH   \n",
       "437           GTA             GA              frameshift_variant   HIGH   \n",
       "438   GAAAAATAAGC     GAAAATAAGC              frameshift_variant   HIGH   \n",
       "627         TGGCC           TGCC              frameshift_variant   HIGH   \n",
       "294             T              C                       stop_lost   HIGH   \n",
       "504             C              T                     stop_gained   HIGH   \n",
       "388         GCCTG         GCCCTG              frameshift_variant   HIGH   \n",
       "525         TAAAC          AAAAT                     stop_gained   HIGH   \n",
       "245         CCCTA          TCCTT                     stop_gained   HIGH   \n",
       "143      GGTCAATA       AGACAACA                       stop_lost   HIGH   \n",
       "277          GGCA           AGCT                     stop_gained   HIGH   \n",
       "347       AAGAAAA        TAGTAAT                     stop_gained   HIGH   \n",
       "528       GTTTTCA         GTTTCA              frameshift_variant   HIGH   \n",
       "654             T              A                       stop_lost   HIGH   \n",
       "292           GTC            ATT                     stop_gained   HIGH   \n",
       "284          TATA           AATT                     stop_gained   HIGH   \n",
       "619    TTTCCCTAAA     CTTCTCTTAA                     stop_gained   HIGH   \n",
       "508         TTTTT          ATTTA                     stop_gained   HIGH   \n",
       "\n",
       "       AFF                           MUT  \n",
       "118    cat                     p.Tyr176*  \n",
       "130    cat                     p.Tyr211*  \n",
       "47     cat                      p.Glu2fs  \n",
       "54     cat                      p.Gln15*  \n",
       "55     cat                      p.Trp16*  \n",
       "59     cat                      p.Gln30*  \n",
       "93     cat                  p.LeuTrp105*  \n",
       "384    cat                     p.Gln209*  \n",
       "331    cat                      p.Ile5fs  \n",
       "337    cat                      p.Gln26*  \n",
       "519    cat               p.CysTyrGly126*  \n",
       "506    cat                   p.ValGln37*  \n",
       "520    cat                     p.Lys136*  \n",
       "523    cat                     p.Trp150*  \n",
       "502    cat                      p.Gln26*  \n",
       "129    cat                     p.Gln210*  \n",
       "191    cat                 p.MetGluLys1?  \n",
       "260    cat                     p.Tyr176*  \n",
       "197    cat                      p.His17*  \n",
       "213    cat                   p.LeuLys45*  \n",
       "200    cat                      p.Gln26*  \n",
       "138  lacI4                    p.ValPhe6*  \n",
       "601    cat                      p.Lys54*  \n",
       "139  lacI4                      p.Lys14*  \n",
       "595    cat                      p.Gln26*  \n",
       "393  lacI4        p.TerArg373LeuGlnext*?  \n",
       "437    cat                     p.Val48fs  \n",
       "438    cat                     p.Asn51fs  \n",
       "627    cat                    p.Ala161fs  \n",
       "294  lacI4              p.Ter373Glnext*?  \n",
       "504    cat                      p.Gln30*  \n",
       "388  lacI4                     p.Val16fs  \n",
       "525    cat                     p.Leu158*  \n",
       "245    cat                     p.Lys136*  \n",
       "143  lacI4  p.GlyGlnTer371ArgGlnGlnext*?  \n",
       "277    cat                  p.TrpGln215*  \n",
       "347    cat                p.LysLysAsn49*  \n",
       "528    cat                    p.Phe171fs  \n",
       "654  lacI4              p.Ter375Lysext*?  \n",
       "292  lacI4                  p.ArgGln363*  \n",
       "284  lacI4                   p.LeuTyr18*  \n",
       "619    cat                  p.ProLys135*  \n",
       "508    cat                   p.PheLeu44*  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect22[effect22['IMPACT']=='HIGH'].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>149</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.449275</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>154</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.612903</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>176</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.724138</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508435</td>\n",
       "      <td>822775</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>GACGC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508436</td>\n",
       "      <td>822787</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.142857</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|synonymous_variant|LOW|soj|MPN688|transcript...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508437</td>\n",
       "      <td>822830</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|missense_variant|MODERATE|soj|MPN688|transcr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508438</td>\n",
       "      <td>822867</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>CTTTC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>508439</td>\n",
       "      <td>822913</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508440 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           POS        REF        ALT          QUAL  TOT  REFN  ALTN ALTS  \\\n",
       "0           70          C          T  4.927350e-15   60    58     2    2   \n",
       "1          149          T          C  0.000000e+00   75    74     1    1   \n",
       "2          150          A          T  2.263910e-15   69    68     1    1   \n",
       "3          154          T          C  0.000000e+00   62    61     1    1   \n",
       "4          176       TAAT       CAAC  3.420820e-15   58    57     1    1   \n",
       "...        ...        ...        ...           ...  ...   ...   ...  ...   \n",
       "508435  822775      AACGT      GACGC  3.631530e-15  171   170     1    1   \n",
       "508436  822787          A          G  1.152910e-15  175   173     2    2   \n",
       "508437  822830          T          G  3.857310e-15  165   164     1    1   \n",
       "508438  822867      TTTTT      CTTTC  0.000000e+00  172   171     1    1   \n",
       "508439  822913  CGCGCGTAT  TGCGTGAAA  0.000000e+00  161   160     1    1   \n",
       "\n",
       "            FRAC        SAMPLE   P  \\\n",
       "0       3.333333       p2_2_S1   2   \n",
       "1       1.333333       p2_2_S1   2   \n",
       "2       1.449275       p2_2_S1   2   \n",
       "3       1.612903       p2_2_S1   2   \n",
       "4       1.724138       p2_2_S1   2   \n",
       "...          ...           ...  ..   \n",
       "508435  0.584795  p3IPTG_A_S1b  18   \n",
       "508436  1.142857  p3IPTG_A_S1b  18   \n",
       "508437  0.606061  p3IPTG_A_S1b  18   \n",
       "508438  0.581395  p3IPTG_A_S1b  18   \n",
       "508439  0.621118  p3IPTG_A_S1b  18   \n",
       "\n",
       "                                                      EFF  \n",
       "0       T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "1       C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "2       T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "3       C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "4       CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...  \n",
       "...                                                   ...  \n",
       "508435  GACGC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "508436  G|synonymous_variant|LOW|soj|MPN688|transcript...  \n",
       "508437  G|missense_variant|MODERATE|soj|MPN688|transcr...  \n",
       "508438  CTTTC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "508439  TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...  \n",
       "\n",
       "[508440 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3bfe1f931ce8464f8512a8a6005ef7b1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f92f451a828>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "effect21['logFRAC'] = np.log2(effect21['FRAC'])\n",
    "sns.violinplot(x='IMPACT', y='FRAC', hue='PASS', data=effect1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>18482</td>\n",
       "      <td>566325</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.469330e-15</td>\n",
       "      <td>205</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.487805</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18481</td>\n",
       "      <td>566311</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>7.046470e-16</td>\n",
       "      <td>198</td>\n",
       "      <td>197</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.505051</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|missense_variant|MODERATE|Gene_565549_566677...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18485</td>\n",
       "      <td>566383</td>\n",
       "      <td>GCC</td>\n",
       "      <td>CCT</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>181</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.552486</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CCT|missense_variant|MODERATE|Gene_565549_5666...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18483</td>\n",
       "      <td>566361</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18476</td>\n",
       "      <td>566244</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.167980e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6416</td>\n",
       "      <td>195428</td>\n",
       "      <td>CAAAAAAAAAAAAAAAGTAAAATAGAAAAGC</td>\n",
       "      <td>CAAAATAAGAAAAAAAGTAAAATAGAAAAGC,CAAAAAAAAAAAAA...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>24</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>1,1,1,6</td>\n",
       "      <td>37.500000</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAAATAAGAAAAAAAGTAAAATAGAAAAGC|upstream_gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17103</td>\n",
       "      <td>528761</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTGAAAGA,GTTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>6,6</td>\n",
       "      <td>42.857143</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTGAAAGA|upstream_gene_variant|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20563</td>\n",
       "      <td>629173</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>GTTTTTTTTTATTTATAAGTTTGAAC,GTTTTTTTTTTTTTTTAGT...</td>\n",
       "      <td>1.383230e-14</td>\n",
       "      <td>30</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>1,2,6,2,2</td>\n",
       "      <td>43.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>GTTTTTTTTTATTTATAAGTTTGAAC|upstream_gene_varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4753</td>\n",
       "      <td>141270</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGC,CAGAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>52</td>\n",
       "      <td>27</td>\n",
       "      <td>25</td>\n",
       "      <td>7,18</td>\n",
       "      <td>48.076923</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGC|upstream_gene_variant|MOD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19735</td>\n",
       "      <td>602077</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.607630e+01</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>A|missense_variant|MODERATE|P02_orf1300|MPN489...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27073 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          POS                              REF  \\\n",
       "18482  566325                                C   \n",
       "18481  566311                                A   \n",
       "18485  566383                              GCC   \n",
       "18483  566361                                A   \n",
       "18476  566244                                A   \n",
       "...       ...                              ...   \n",
       "6416   195428  CAAAAAAAAAAAAAAAGTAAAATAGAAAAGC   \n",
       "17103  528761         GTTTTTTTTTTTTTTTTTGAAAGA   \n",
       "20563  629173       GTTTTTTTTTTTTTTTTAGTTTGAAC   \n",
       "4753   141270           CAGAGAGAGAGAGAGAGAGAGC   \n",
       "19735  602077                                T   \n",
       "\n",
       "                                                     ALT          QUAL  TOT  \\\n",
       "18482                                                  T  1.469330e-15  205   \n",
       "18481                                                  G  7.046470e-16  198   \n",
       "18485                                                CCT  0.000000e+00  181   \n",
       "18483                                                  G  0.000000e+00  172   \n",
       "18476                                                  G  1.167980e-15  171   \n",
       "...                                                  ...           ...  ...   \n",
       "6416   CAAAATAAGAAAAAAAGTAAAATAGAAAAGC,CAAAAAAAAAAAAA...  0.000000e+00   24   \n",
       "17103  GTTTTTTTTTTTTTTTTGAAAGA,GTTTTTTTTTTTTTTTTTTGAAAGA  0.000000e+00   28   \n",
       "20563  GTTTTTTTTTATTTATAAGTTTGAAC,GTTTTTTTTTTTTTTTAGT...  1.383230e-14   30   \n",
       "4753       CAGAGAGAGAGAGAGAGAGC,CAGAGAGAGAGAGAGAGAGAGAGC  0.000000e+00   52   \n",
       "19735                                                  A  1.607630e+01    1   \n",
       "\n",
       "       REFN  ALTN       ALTS        FRAC   SAMPLE  P  \\\n",
       "18482   204     1          1    0.487805  p2_2_S1  2   \n",
       "18481   197     1          1    0.505051  p2_2_S1  2   \n",
       "18485   180     1          1    0.552486  p2_2_S1  2   \n",
       "18483   171     1          1    0.581395  p2_2_S1  2   \n",
       "18476   170     1          1    0.584795  p2_2_S1  2   \n",
       "...     ...   ...        ...         ...      ... ..   \n",
       "6416     15     9    1,1,1,6   37.500000  p2_2_S1  2   \n",
       "17103    16    12        6,6   42.857143  p2_2_S1  2   \n",
       "20563    17    13  1,2,6,2,2   43.333333  p2_2_S1  2   \n",
       "4753     27    25       7,18   48.076923  p2_2_S1  2   \n",
       "19735     0     1          1  100.000000  p2_2_S1  2   \n",
       "\n",
       "                                                     EFF  \n",
       "18482  T|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "18481  G|missense_variant|MODERATE|Gene_565549_566677...  \n",
       "18485  CCT|missense_variant|MODERATE|Gene_565549_5666...  \n",
       "18483  G|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "18476  G|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "...                                                  ...  \n",
       "6416   CAAAATAAGAAAAAAAGTAAAATAGAAAAGC|upstream_gene_...  \n",
       "17103  GTTTTTTTTTTTTTTTTGAAAGA|upstream_gene_variant|...  \n",
       "20563  GTTTTTTTTTATTTATAAGTTTGAAC|upstream_gene_varia...  \n",
       "4753   CAGAGAGAGAGAGAGAGAGC|upstream_gene_variant|MOD...  \n",
       "19735  A|missense_variant|MODERATE|P02_orf1300|MPN489...  \n",
       "\n",
       "[27073 rows x 12 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls1[snpcalls1['SAMPLE']=='p2_2_S1'].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'annotations' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-d335a4bfad19>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msimplify_effect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnpcalls1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-46-bf2fb86f217b>\u001b[0m in \u001b[0;36msimplify_effect\u001b[0;34m(df, positions, genome, modify_original, impact, mintot, minrefn, minaltn)\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mannotations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcorrespondance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpositions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m         \u001b[0mannotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mannotations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqualifiers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'label'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'label'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqualifiers\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'annotations' referenced before assignment"
     ]
    }
   ],
   "source": [
    "simplify_effect(snpcalls1, genome=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c98a5e27e0e4195a0e684646f587140",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure\n",
    "\n",
    "\n",
    "def plot_analysis(df):\n",
    "    repro = [k for k, v in Counter(list(df.POS)).items() if v>=2]\n",
    "    plt.figure()\n",
    "    plt.subplot(3,3,1)\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(posE)) & (df['POS'].isin(repro))], label='E')\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(posN)) & (df['POS'].isin(repro))], label='NE')\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(cas1)) & (df['POS'].isin(repro))], label='C')\n",
    "\n",
    "\n",
    "plot_analysis(snpcalls1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Substituion rate per base comparing cassette versus other distributions\n",
    "\n",
    "To evaluate the selection in the cassette we compare the rate of substitutions per base occuring within the cassette, a set of essential and non-essential genes and with the general distribution observed at genome level. In this case we do not care about the fraction each variant is found but how many times a variant is found in the cassette versus the other distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show the ratio of mutation in cassette versus general\n",
    "def mutation_rate(snpdf, segment, mintot=5, minrefn=3, minaltn=2):\n",
    "    _df = snpdf[(snpdf['TOT']>=mintot) & (snpdf['REFN']>=minrefn) & (snpdf['ALTN']>=minaltn)].copy()    \n",
    "    rs = {}\n",
    "    for sample in set(_df.SAMPLE):\n",
    "        total_len = 816394\n",
    "        _df2 = _df[_df['SAMPLE']==sample].copy()\n",
    "        NE = _df2[_df2['POS'].isin(posN)]\n",
    "        ES = _df2[_df2['POS'].isin(posE)]\n",
    "        TG = _df2[~_df2['POS'].isin(segment)]\n",
    "        CS = _df2[_df2['POS'].isin(segment)]\n",
    "        #total_len+=len(segment)\n",
    "        rs[sample] = [sample, sample.split('_')[0],\n",
    "                      int(sample.split('_')[0].replace('p', '').replace('3IPTG', '18')),\n",
    "                      TG.shape[0], ES.shape[0], NE.shape[0], CS.shape[0], \n",
    "                      sum(TG['FRAC']), sum(ES['FRAC']), sum(NE['FRAC']), sum(CS['FRAC'])]\n",
    "    jj = pd.DataFrame.from_dict(rs, orient='index')\n",
    "    jj.columns = ['SAMPLE', 'COND', 'P', 'T', 'E', 'N', 'C', 'TF', 'EF', 'NF', 'CF']\n",
    "    jj['T%'] = 100*jj['T']/(total_len)  \n",
    "    jj['E%'] = 100*jj['E']/len(posE)\n",
    "    jj['N%'] = 100*jj['N']/len(posN)\n",
    "    jj['C%'] = 100*jj['C']/len(segment)\n",
    "    return jj\n",
    "\n",
    "def tp(df, order):\n",
    "    rs = {}\n",
    "    c=0\n",
    "    for row, col in df.iterrows():\n",
    "        rs[c] = [col['COND'],'Essential', col['E%'], col['EF']]\n",
    "        rs[c+1] = [col['COND'],'Non-essential', col['N%'], col['NF']]\n",
    "        rs[c+2] = [col['COND'],'Chromosome', col['T%'], col['TF']]\n",
    "        rs[c+3] = [col['COND'],'Cassette', col['C%'], col['CF']]\n",
    "        c+=4\n",
    "    a = pd.DataFrame.from_dict(rs, orient='index')\n",
    "    a.columns = ['Condition', 'Loci', 'Variation per base [%]', 'Accumulated Fraction [%]']\n",
    "    return orderdf(a, ordered_classes=order, col='Condition')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutrate11 = mutation_rate(snpcalls1, segment=cas1)\n",
    "mutrate21 = mutation_rate(snpcalls2, segment=cas1)\n",
    "mutrate22 = mutation_rate(snpcalls2, segment=cas2)\n",
    "mutrate212 = mutation_rate(snpcalls2, segment=cas12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "25d6f7201093424d901273e1957b28ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.close('all')\n",
    "plt.figure(figsize=(15, 3))\n",
    "c = 1\n",
    "for text, df in zip(['C5 Par', 'C20 Par', 'C20 3b'], [mutrate11, mutrate21, mutrate22]):\n",
    "    plt.subplot(1,3,c)\n",
    "    if text=='C5 Par':\n",
    "        sns.barplot(x='Condition', y='Variation per base [%]', hue='Loci', palette='mako', data=tp(df, ['p2','p3','p15', 'p3IPTG']))\n",
    "    else:\n",
    "        sns.barplot(x='Condition', y='Variation per base [%]', hue='Loci', palette='mako', data=tp(df, ['p2','p3','p15']))\n",
    "    plt.ylim(0, 15)\n",
    "    c+=1\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dnds_functions import evolution_of_sample, dnds\n",
    "ttt = evolution_of_sample(effect1[effect1['IMPACT']=='HIGH'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "effectC= simplify_effect(snpcalls1, cas1, minaltn=3)\n",
    "effectE= simplify_effect(snpcalls1, posE, minaltn=3)\n",
    "effectN= simplify_effect(snpcalls1, posN, minaltn=3)\n",
    "\n",
    "tttC = evolution_of_sample(effectC)\n",
    "tttE = evolution_of_sample(effectE)\n",
    "tttN = evolution_of_sample(effectN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'p15_F_S1': [0.015,\n",
       "  0.011,\n",
       "  1.3636363636363638,\n",
       "  0.015,\n",
       "  0.011,\n",
       "  1.3636363636363638],\n",
       " 'p3IPTG_A_S1': [0.418,\n",
       "  0.265,\n",
       "  1.5773584905660376,\n",
       "  0.612,\n",
       "  0.327,\n",
       "  1.8715596330275228],\n",
       " 'p3_D_S3': [0.012, 0.0, '+', 0.012, 0.0, '+'],\n",
       " 'p2_2_S1': [0, 0, '/', 0, 0, '/'],\n",
       " 'p3IPTG_A_S1b': [0, 0, '/', 0, 0, '/'],\n",
       " 'p15_F_S5': [0, 0, '/', 0, 0, '/'],\n",
       " 'p2_2_S3': [0.009000000000000001,\n",
       "  0.022,\n",
       "  0.40909090909090917,\n",
       "  0.009000000000000001,\n",
       "  0.022,\n",
       "  0.40909090909090917],\n",
       " 'p2_2_S2': [0.006, 0.0, '+', 0.006, 0.0, '+'],\n",
       " 'p3_D_S2': [0.03,\n",
       "  0.06599999999999999,\n",
       "  0.4545454545454546,\n",
       "  0.03,\n",
       "  0.06599999999999999,\n",
       "  0.4545454545454546]}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tttE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "155ba62f89ef40138481c9a168d64db1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f92ec6b1be0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "x, y = [[], []]\n",
    "for k, v in tttC.items():\n",
    "    if type(v[-1])==float and type(tttE[k][-1])==float:\n",
    "        x.append(tttE[k][-1])\n",
    "        y.append(v[-1])\n",
    "plt.scatter(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.subplot(2,2,1)\n",
    "sns.countplot(x='SAMPLE', data=snpcalls1)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Variant count per sample')\n",
    "\n",
    "plt.subplot(2,2,2)\n",
    "sns.boxplot(x='SAMPLE', y='FRAC', data=snpcalls1)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Fraction distribution per sample')\n",
    "\n",
    "plt.subplot(2,2,3)\n",
    "sns.countplot(x='SAMPLE', data=snpcalls2)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Variant count per sample')\n",
    "\n",
    "plt.subplot(2,2,4)\n",
    "sns.boxplot(x='SAMPLE', y='FRAC', data=snpcalls2)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Fraction distribution per sample')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "General exploration of the most representative variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "plt.figure(figsize=(15, 10))\n",
    "plt.subplot(3,1,1)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls1[(snpcalls1['FRAC']>=5) & (snpcalls1['TOT']>=10) & (snpcalls1['P']<=15) & (snpcalls1['FRAC']<100)])\n",
    "plt.axvspan(565510, 572114, color='gray', alpha=0.2)\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.subplot(3,1,2)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls1[(snpcalls1['FRAC']>=5) & (snpcalls1['TOT']>=10) & (snpcalls1['P']>15) & (snpcalls1['FRAC']<100)])\n",
    "plt.axvspan(565510, 572114, color='gray', alpha=0.2)\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.subplot(3,1,3)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls2[(snpcalls2['FRAC']>=5) & (snpcalls2['TOT']>=10) & (snpcalls2['FRAC']<100)])\n",
    "plt.axvspan(cass_par[0], cass_par[1], color='gray', alpha=0.2)\n",
    "plt.axvspan(cass_3b[0], cass_3b[1], color='gray', alpha=0.2)\n",
    "plt.xlabel('genome position [bp]')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.savefig('./figures/supfig1.svg')\n",
    "plt.savefig('./figures/supfig1.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate final snp working data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'evaluate_cassete_mut' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-183126911c41>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mcass_3b\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m780850\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m787445\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mall_f1_par\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mevaluate_cassete_mut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnpcalls1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msten\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcass_par\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mall_f2_par\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mevaluate_cassete_mut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnpcalls2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msten\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcass_par\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mall_f2_3b\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mevaluate_cassete_mut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnpcalls2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msten\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcass_3b\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'evaluate_cassete_mut' is not defined"
     ]
    }
   ],
   "source": [
    "cass_par = [565510, 572124]\n",
    "cass_3b = [780850, 787445]\n",
    "\n",
    "all_f1_par = evaluate_cassete_mut(snpcalls1, sten=cass_par)\n",
    "all_f2_par = evaluate_cassete_mut(snpcalls2, sten=cass_par)\n",
    "all_f2_3b = evaluate_cassete_mut(snpcalls2, sten=cass_3b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f2_3b' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-d5a914627671>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mall_f2_3b\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f2_3b' is not defined"
     ]
    }
   ],
   "source": [
    "all_f2_3b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selection of variants\n",
    "In this case we evaluate the fraction of each variant and how it is fixed along passages. As not all the samples are directly related, we consider the induced sample (IPTG) as 18. In this sense, if a mutation found in passages 2, 3 or 15 is found in 18 with a higher rate it implies that this variants are able to confere the capability to surpass the killswitch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_evolution(df):\n",
    "    positions = set(df[df['FRAC']>1].POS)\n",
    "    plt.figure()\n",
    "    plt.subplot(1,2,1)\n",
    "    sns.lineplot(x='P', y='FRAC', hue='POS', data=df[df['POS'].isin(positions)])\n",
    "    plt.subplot(1,2,2)\n",
    "    sns.boxplot(x='P', y='FRAC', data=df[df['POS'].isin(positions)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f1_par' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-44813b854db5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_f1_par\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f1_par' is not defined"
     ]
    }
   ],
   "source": [
    "plot_evolution(all_f1_par)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f2_par' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-58206eee6826>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_f2_par\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f2_par' is not defined"
     ]
    }
   ],
   "source": [
    "plot_evolution(all_f2_par)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f2_3b' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-346dcfda44f1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_f2_3b\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f2_3b' is not defined"
     ]
    }
   ],
   "source": [
    "plot_evolution(all_f2_3b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'u' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-62aa5fc51987>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgenome_seq2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mu\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_genome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./data/C20_refSeq.gb'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'u' is not defined"
     ]
    }
   ],
   "source": [
    "genome_seq2 = u.load_genome('./data/C20_refSeq.gb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f2_3b' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-4b5bc512a260>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mall_f2_3b\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mall_f2_3b\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'FRAC'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f2_3b' is not defined"
     ]
    }
   ],
   "source": [
    "all_f2_3b[all_f2_3b['FRAC']>=100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_seq2' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-cedf330c6194>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m'CCCCCCCGGATCCCTCG'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_seq2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_seq2' is not defined"
     ]
    }
   ],
   "source": [
    "'CCCCCCCGGATCCCTCG' in genome_seq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_f1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-26-683cd8d5997a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mall_f1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_f1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'FRAC'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>=\u001b[0m\u001b[0;36m2.5\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mall_f1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'P'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m18\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mselected\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'all_f1' is not defined"
     ]
    }
   ],
   "source": [
    "selected = all_f1[(all_f1['FRAC']>=2.5) & (all_f1['P']==18)]\n",
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-27-5cb9f18890e0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mselected\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'FRAC'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "sum(selected['FRAC'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The mutations found affect Cas9B and LacI4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-e8ee7eb5b1fa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mselected\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'EFF'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "[i.split(',') for i in list(selected['EFF'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'annotations' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-15768c84d2bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mannotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'annotations' is not defined"
     ]
    }
   ],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376.6666666666667"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(566678-565549+1)/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'LacI4': [565549, 566678, '-'],\n",
       " 'pS': [566678, 566721, '-'],\n",
       " 'Par cassette': [566747, 567674, '-'],\n",
       " 'cas9B': [567711, 571818, '-'],\n",
       " 'pG64': [571818, 571934, '-'],\n",
       " 'gRNA2': [571942, 572044, '-'],\n",
       " 'protospacer12 (10 targets)': [572024, 572044, '-'],\n",
       " 'p438': [572044, 572066, '-'],\n",
       " 'IR-OR': [572088, 572114, '-'],\n",
       " 'MTn insertion point': [572113, 572124, '-']}"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376.3333333333333\n",
      "1369.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/Bio/Seq.py:2309: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "genes_affected = {}\n",
    "for k, v in annotations.items():\n",
    "    if k in ['cas9B', 'LacI4']:\n",
    "        if k=='cas9B':\n",
    "            seq = Seq(genome_C5[v[0]:v[1]])\n",
    "        else:\n",
    "            seq = Seq(genome_C5[v[0]:v[1]])\n",
    "        print(len(seq)/3)\n",
    "        rvc = seq.reverse_complement()\n",
    "        mrn = rvc.transcribe()\n",
    "        prt = mrn.translate(table=4)\n",
    "        genes_affected[k] = [seq, rvc, mrn, prt, len(prt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('RLESGQ*R*V', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_affected['LacI4'][3][-10:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LacI4 has extra bases in the annotation, that makes it to be non-codonic. I will consider the first stop found (move the end 10 bases)..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'annotations' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-84fcf2bf1dbc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeq\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mgenes_affected\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'cas9B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LacI4'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'cas9B'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'annotations' is not defined"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "genes_affected = {}\n",
    "for k, v in annotations.items():\n",
    "    if k in ['cas9B', 'LacI4']:\n",
    "        if k=='cas9B':\n",
    "            seq = Seq(genome_seq[v[0]:v[1]])\n",
    "        else:\n",
    "            seq = Seq(genome_seq[v[0]+10:v[1]])\n",
    "        print(len(seq)/3)\n",
    "        rvc = seq.reverse_complement()\n",
    "        mrn = rvc.transcribe()\n",
    "        prt = mrn.translate(table=4)\n",
    "        genes_affected[k] = [seq, rvc, mrn, prt, len(prt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'LacI4'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-2cf0e14076cd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgenes_affected\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'LacI4'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'LacI4'"
     ]
    }
   ],
   "source": [
    "genes_affected['LacI4'][3][-10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluate effect of the selected mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>149</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1.449275</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>154</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1.612903</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>176</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1.724138</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41166</td>\n",
       "      <td>822775</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>GACGC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41167</td>\n",
       "      <td>822787</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>1.142857</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|synonymous_variant|LOW|soj|MPN688|transcript...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41168</td>\n",
       "      <td>822830</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|missense_variant|MODERATE|soj|MPN688|transcr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41169</td>\n",
       "      <td>822867</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>CTTTC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41170</td>\n",
       "      <td>822913</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508440 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          POS        REF        ALT          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "0          70          C          T  4.927350e-15   60    58     2  3.333333   \n",
       "1         149          T          C  0.000000e+00   75    74     1  1.333333   \n",
       "2         150          A          T  2.263910e-15   69    68     1  1.449275   \n",
       "3         154          T          C  0.000000e+00   62    61     1  1.612903   \n",
       "4         176       TAAT       CAAC  3.420820e-15   58    57     1  1.724138   \n",
       "...       ...        ...        ...           ...  ...   ...   ...       ...   \n",
       "41166  822775      AACGT      GACGC  3.631530e-15  171   170     1  0.584795   \n",
       "41167  822787          A          G  1.152910e-15  175   173     2  1.142857   \n",
       "41168  822830          T          G  3.857310e-15  165   164     1  0.606061   \n",
       "41169  822867      TTTTT      CTTTC  0.000000e+00  172   171     1  0.581395   \n",
       "41170  822913  CGCGCGTAT  TGCGTGAAA  0.000000e+00  161   160     1  0.621118   \n",
       "\n",
       "             SAMPLE   P                                                EFF  \n",
       "0           p2_2_S1   2  T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "1           p2_2_S1   2  C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "2           p2_2_S1   2  T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "3           p2_2_S1   2  C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "4           p2_2_S1   2  CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...  \n",
       "...             ...  ..                                                ...  \n",
       "41166  p3IPTG_A_S1b  18  GACGC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "41167  p3IPTG_A_S1b  18  G|synonymous_variant|LOW|soj|MPN688|transcript...  \n",
       "41168  p3IPTG_A_S1b  18  G|missense_variant|MODERATE|soj|MPN688|transcr...  \n",
       "41169  p3IPTG_A_S1b  18  CTTTC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "41170  p3IPTG_A_S1b  18  TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...  \n",
       "\n",
       "[508440 rows x 11 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-96ce78c8031b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mselected\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0meff\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEFF\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mjjj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0meff\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'|'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-184-19ff2ad0f5a8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected_ann' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-38-28fde1192669>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected_ann\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected_ann' is not defined"
     ]
    }
   ],
   "source": [
    "selected_ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected_ann' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-e70eb41dc56f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected_ann\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mselected_ann\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'IMPACT'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'HIGH'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected_ann' is not defined"
     ]
    }
   ],
   "source": [
    "selected_ann[selected_ann['IMPACT']=='HIGH']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation rate study\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-d63c0ec7e1ed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmelt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'P'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'%'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "pd.melt(mutrate1.sort_values('P')[[i for i in mutrate1.columns if '%' in i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0f22cbb724f46aebea1cae18e8be94f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-45-05ad25cb1d90>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mboxplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'N%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutrate2\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'N%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.boxplot([mutrate1['N%'], mutrate2['N%']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0774f1234ea48ddac4ae853136b8b18",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-46-c03ce5ec3171>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'OUT%'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IN%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'grey'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlinestyle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'--'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatterplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     plt.text(x=mutrate1[x][i]+0.2,y=mutrate1[y][i]+0.3,s=mutrate1.index[i],\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "plt.close('all')\n",
    "plt.figure()\n",
    "x, y = ['OUT%', 'IN%']\n",
    "plt.plot([0,16],[0,16], c='grey', linestyle='--')\n",
    "sns.scatterplot(x=x, y=y, data=mutrate1)\n",
    "for i in range(mutrate1.shape[0]):\n",
    "    plt.text(x=mutrate1[x][i]+0.2,y=mutrate1[y][i]+0.3,s=mutrate1.index[i],\n",
    "             fontdict=dict(size=8))\n",
    "sns.scatterplot(x=x, y=y, data=mutrate2)\n",
    "c=0\n",
    "for i in range(mutrate2.shape[0]):\n",
    "\n",
    "    if c==0:\n",
    "        plt.text(x=mutrate2[x][i]+0.2,y=mutrate2[y][i]+0.4,s=mutrate2.index[i],\n",
    "                 fontdict=dict(size=10))\n",
    "        c==1\n",
    "    else:\n",
    "        plt.text(x=mutrate2[x][i]+0.2,y=mutrate2[y][i]+0.2,s=mutrate2.index[i],\n",
    "                 fontdict=dict(size=8))\n",
    "        c==0\n",
    "plt.xlabel('Genome Variant Rate [%]')\n",
    "plt.ylabel('Cassette Variant Rate [%]')\n",
    "plt.xlim(0,16)\n",
    "plt.ylim(0,16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mutrate' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-8c5c8001c3c9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mwrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCounter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     wrs[p] = wilcoxon(x=list(mutrate[mutrate['P']==p]['OUT%']),\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate' is not defined"
     ]
    }
   ],
   "source": [
    "from scipy.stats import wilcoxon, ttest_rel\n",
    "from collections import Counter\n",
    "\n",
    "wrs = {}\n",
    "trs = {}\n",
    "ns = dict(Counter(mutrate.P))\n",
    "for p in set(mutrate.P):\n",
    "    wrs[p] = wilcoxon(x=list(mutrate[mutrate['P']==p]['OUT%']),\n",
    "                      y=list(mutrate[mutrate['P']==p]['IN%']))[1]\n",
    "    trs[p] = ttest_rel(list(mutrate[mutrate['P']==p]['OUT%']),\n",
    "                       list(mutrate[mutrate['P']==p]['IN%']))[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ns' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-5ec18e92d2c6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'ns' is not defined"
     ]
    }
   ],
   "source": [
    "ns, wrs, trs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "null hypothesis (same mut rate) cannot be rejected at a confidence level of 5%, only at early passages 2 and 3 seems to be limited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snpcalls50' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-bf7c51358a6e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msnpcalls50\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'snpcalls50' is not defined"
     ]
    }
   ],
   "source": [
    "snpcalls50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot variations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[565510, 572124]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_par"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-51-19ff2ad0f5a8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "\n",
    "\n",
    "def plot_variations(genbank_file, coords, variations):\n",
    "    record = SeqIO.read(genbank_file, \"genbank\")\n",
    "\n",
    "    gd_diagram = GenomeDiagram.Diagram(record.id)\n",
    "    gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "    for feature in record.features:\n",
    "        if feature.location.start>=coords[0] and feature.location.start<=coords[1] and feature.type!='Polymorphism':\n",
    "            print(feature.type, feature.qualifiers['label'], feature.location)\n",
    "            feature.location.strand=-1\n",
    "            #if feature.type != \"gene\":\n",
    "                # Exclude this feature\n",
    "            #    continue\n",
    "            #if len(gd_feature_set) % 2 == 0:\n",
    "            #    color = colors.blue\n",
    "            if feature.type=='CDS':\n",
    "                color = colors.lightblue\n",
    "                gd_feature_set.add_feature(\n",
    "                    feature, sigil=\"ARROW\", arrowshaft_height=1.0, color=color, height=0.2, label=True, label_size=14, label_angle=0\n",
    "                )\n",
    "            # else:\n",
    "            #     color = colors.lightblue\n",
    "            #     gd_feature_set.add_feature(\n",
    "            #         feature, sigil=\"BOX\", color=color, height=0.1, label=True, label_size=14, label_angle=10)\n",
    "    colordic = {'stop': colors.firebrick, 'frameshift':colors.red, 'missense':colors.orange}\n",
    "    done = []\n",
    "\n",
    "    for pos, eff in zip(variations.POS, variations.EFF):\n",
    "        if pos not in done:\n",
    "            effi = eff.split('|')[1].split('_')[0]\n",
    "            color = colordic[effi]\n",
    "            feature = SeqFeature(FeatureLocation(pos, pos+1))\n",
    "            gd_feature_set.add_feature(\n",
    "                    feature,\n",
    "                    color=color,\n",
    "                    name=' '+effi,\n",
    "                    label=True,\n",
    "                    label_size=15,\n",
    "                    size=20\n",
    "                    #label_color=color,\n",
    "                )\n",
    "            done.append(pos)\n",
    "            for i in range(pos-30, pos+30):\n",
    "                done.append(i)\n",
    "    gd_diagram.draw(format=\"linear\", pagesize=\"A5\", fragments=1, start=565510, end=572124)\n",
    "    gd_diagram.write(\"plasmid_linear_nice.svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-53-3c5b915f50d3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_variations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./data/C5_refSeq.gb'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcoords\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m565549\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m571818\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariations\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mselected\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "plot_variations('./data/C5_refSeq.gb', coords=[565549,571818], variations=selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 637,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "268.6666666666668\n"
     ]
    }
   ],
   "source": [
    "raw_seq, new_seq = get_new_sequence(seq_record5, 'LacI4', pos=[565576, 'AGGC', 'GGGT'])\n",
    "yyy = dnds(raw_seq, new_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 648,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2140</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>571815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>674</td>\n",
       "      <td>668</td>\n",
       "      <td>6</td>\n",
       "      <td>0.898204</td>\n",
       "      <td>CCAGATGTAA</td>\n",
       "      <td>TCTGTTGTTT</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.LeuAsp1?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE  PASS     POS  QUAL  TOT  REFN  ALTN      FRAC         REF  \\\n",
       "2140  p3_D_S2     3  571815   0.0  674   668     6  0.898204  CCAGATGTAA   \n",
       "\n",
       "             ALT         EFF IMPACT    AFF         MUT  \n",
       "2140  TCTGTTGTTT  start_lost   HIGH  cas9B  p.LeuAsp1?  "
      ]
     },
     "execution_count": 648,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect1[effect1['POS']==571815]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 656,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p3IPTG_A_S1 2.404754044239023\n",
      "p3_D_S2 1.7427616926503313\n",
      "p15_F_S1 2.3935483870967738\n",
      "p15_F_S5 3.4130506790327924\n",
      "p3IPTG_A_S1b 2.6608581600149894\n",
      "p3_D_S3 2.117820324005891\n",
      "p2_2_S3 2.3640178337267255\n",
      "p2_2_S1 2.127371273712737\n",
      "p2_2_S2 1.7000000000000002\n"
     ]
    }
   ],
   "source": [
    "for k, v in ttt.items():\n",
    "    print(k, v[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 620,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-56-9a8f373c509c>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-56-9a8f373c509c>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    df = pd.read_csv(, header=60, sep='\\t')\u001b[0m\n\u001b[0m                     ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(, header=60, sep='\\t')\n",
    "df2 = df[(df['INFO'].str.contains('CIGAR=1X;'))].sample(n=150, random_state=1).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "Missing parentheses in call to 'print'. Did you mean print(args.query, chromosome, pn, ps, pns, dn, ds, dns)? (<ipython-input-57-012244273d6f>, line 18)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-57-012244273d6f>\"\u001b[0;36m, line \u001b[0;32m18\u001b[0m\n\u001b[0;31m    print args.query, chromosome, pn, ps, pns, dn, ds, dns\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m Missing parentheses in call to 'print'. Did you mean print(args.query, chromosome, pn, ps, pns, dn, ds, dns)?\n"
     ]
    }
   ],
   "source": [
    "pn, ps, pns, dn, ds, dns = 0, 0, 0, 0, 0, 0\n",
    "for item in sorted(output[chromosome]):\n",
    "    raw_seq, new_seq = get_new_sequence(args.reference, chromosome, output[chromosome][item][0].qualifiers[\n",
    "        'locus_tag'][0], output[chromosome][item][1])\n",
    "    newpn, newps, newdn, newds = dnds(raw_seq, new_seq)\n",
    "    pn += newpn\n",
    "    ps += newps\n",
    "    dn += newdn\n",
    "    ds += newds\n",
    "if pn == 0 and ps == 0:\n",
    "    pns, dns = '/', '/'\n",
    "elif ps == 0:\n",
    "    pns, dns = '+', '+'\n",
    "elif pn == 0:\n",
    "    pns, dns = '-', '-'\n",
    "else:\n",
    "    pns, dns = pn / ps, dn / ds\n",
    "print args.query, chromosome, pn, ps, pns, dn, ds, dns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-58-9fd2afaa1715>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-58-9fd2afaa1715>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    for records in SeqIO.parse(, \"genbank\"):\u001b[0m\n\u001b[0m                               ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "for records in SeqIO.parse(, \"genbank\"):\n",
    "    print(/)\n",
    "    raw_seq, new_seq = get_new_sequence('./data/C5_refSeq.gb', 'C5', output[chromosome][item][0].qualifiers[\n",
    "        'locus_tag'][0], output[chromosome][item][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
